<!doctype html><html lang=en><head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><link rel=apple-touch-icon sizes=180x180 href=/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/favicon-16x16.png><link rel=manifest href=/site.webmanifest><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><style>:root{--accent-color:#FF4D4D}</style><title>ATAC-seq Insert Size Plotting</title><meta name=description content="Genomics | Bioinformatics | Epigenetics"><meta name=keywords content='blog,gokarna,hugo,data science,biology,genomics,bioinformatics,unix,text manipulation,ATAC-seq,samtools'><meta property="og:url" content="http://localhost:1313/posts/2017-02-10-atac-seq-insert-size-plotting/"><meta property="og:type" content="website"><meta property="og:title" content="ATAC-seq Insert Size Plotting"><meta property="og:description" content="Genomics | Bioinformatics | Epigenetics"><meta property="og:image" content="/images/profile_avatar.png"><meta name=twitter:card content="summary_large_image"><meta name=twitter:title content="ATAC-seq Insert Size Plotting"><meta name=twitter:description content="Genomics | Bioinformatics | Epigenetics"><meta property="twitter:domain" content="http://localhost:1313/posts/2017-02-10-atac-seq-insert-size-plotting/"><meta property="twitter:url" content="http://localhost:1313/posts/2017-02-10-atac-seq-insert-size-plotting/"><meta name=twitter:image content="/images/profile_avatar.png"><link rel=canonical href=http://localhost:1313/posts/2017-02-10-atac-seq-insert-size-plotting/><link rel=stylesheet type=text/css href=/css/normalize.min.css media=print onload='this.media="all"'><link rel=stylesheet type=text/css href=/css/main.css><link disabled id=dark-theme rel=stylesheet href=/css/dark.css><script src=/js/svg-injector.min.js></script><script src=/js/feather-icons.min.js></script><script src=/js/main.js></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css integrity=sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ crossorigin=anonymous><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js integrity=sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB crossorigin=anonymous></script><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js integrity=sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl crossorigin=anonymous onload=renderMathInElement(document.body)></script><script>document.addEventListener("DOMContentLoaded",function(){renderMathInElement(document.body,{delimiters:[{left:"$$",right:"$$",display:!0},{left:"$",right:"$",display:!1}]})})</script></head><body><script type=text/javascript>setThemeByUserPref()</script><header class=header><nav class=header-nav><div class=avatar><a href=http://localhost:1313/><img src=http://localhost:1313//images/profile_avatar.png alt=avatar></a></div><div class=nav-title><a class=nav-brand href=http://localhost:1313/>Not A Rocket Scientist</a></div><div class=nav-links><div class=nav-link><a href=http://localhost:1313/><span data-feather=home></span> Home</a></div><div class=nav-link><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></div><div class=nav-link><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></div><div class=nav-link><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></div><div class=nav-link><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></div><div class=nav-link><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></div><div class=nav-link><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></div><span class=nav-icons-divider></span><div class="nav-link dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only></span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></div><div class=nav-link id=hamburger-menu-toggle><span id=hamburger-menu-toggle-screen-reader-target class=sr-only>menu</span>
<a><span data-feather=menu></span></a></div><ul class="nav-hamburger-list visibility-hidden"><li class=nav-item><a href=http://localhost:1313/><span data-feather=home></span> Home</a></li><li class=nav-item><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></li><li class=nav-item><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></li><li class=nav-item><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></li><li class=nav-item><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></li><li class=nav-item><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></li><li class=nav-item><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></li><li class="nav-item dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only>theme</span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></li></ul></div></nav></header><main id=content><div class="post container"><div class=post-header-section><h1>ATAC-seq Insert Size Plotting</h1><small role=doc-subtitle></small><p class=post-date>February 10, 2017</p><ul class=post-tags><li class=post-tag><a href=http://localhost:1313/tags/bioinformatics>bioinformatics</a></li><li class=post-tag><a href=http://localhost:1313/tags/unix>unix</a></li><li class=post-tag><a href=http://localhost:1313/tags/text-manipulation>text manipulation</a></li><li class=post-tag><a href=http://localhost:1313/tags/atac-seq>ATAC-seq</a></li><li class=post-tag><a href=http://localhost:1313/tags/samtools>samtools</a></li></ul></div><div class=post-content><p><p>ATAC-seq is THE simplest genomic protocol to profile open chromatin regions in the genome. The original publication (<strong><a href=http://www.nature.com/nmeth/journal/v10/n12/full/nmeth.2688.html>Nature Methods 10, 1213-1218</a></strong>) has been cited 441 times at this moment of writing (1,073 times at this moment of re-posting). It has become almost a routine for a lab that is interested in transcriptional regulation.</p><p>One common QC for the data is to plot the fragment size density of your libraries. The successful construction of a ATAC library requires a proper pair of Tn5 transposase cutting events at the ends of DNA. In the nucleosome-free open chromatin regions, many molecules of Tn5 can kick in and chop the DNA into small pieces; around nucleosome-occupied regions, Tn5 can only access the linker regions. Therefore, in a normal ATAC-seq library, you should expect to see a sharp peak at the &lt;100 bp region (open chromatin), and a peak at ~200bp region (mono-nucleosome), and other larger peaks (multi-nucleosomes). Examples from one of my data:</p><p>Regular scale:</p><p><img src=/images/000/atac_isize_regular.png alt></p><p>Log scale:</p><p><img src=/images/000/atac_isize_log.png alt></p><p>This clear nucleosome phasing pattern indicates a good quality of the experiment.</p><p>Different people probably have different ways of plotting this using different codes, but making this plot can be simply achieved by a combination of one line of unix commands and Excel. You donâ€™t need some complicated scripts to do it at all.</p><p>The 9th column of sam/bam file contains the fragment length information of the paired reads. Therefore, you can count them by:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>samtools view ATAC_f2q30_sorted.bam | awk <span style=color:#98c379>&#39;$9&gt;0&#39;</span> | cut -f <span style=color:#d19a66>9</span> | sort | uniq -c | sort -b -k2,2n | sed -e <span style=color:#98c379>&#39;s/^[ \t]*//&#39;</span> &gt; fragment_length_count.txt
</span></span></code></pre></div><p>This is what it does:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>samtools view ATAC_f2q30_sorted.bam | <span style=color:#7f848e>#stream reads</span>
</span></span><span style=display:flex><span> 
</span></span><span style=display:flex><span>awk <span style=color:#98c379>&#39;$9&gt;0&#39;</span> | <span style=color:#7f848e>#output reads with positive fragment length, i.e. left-most reads</span>
</span></span><span style=display:flex><span> 
</span></span><span style=display:flex><span>cut -f <span style=color:#d19a66>9</span> | <span style=color:#7f848e>#get the fragment length</span>
</span></span><span style=display:flex><span> 
</span></span><span style=display:flex><span>sort | uniq -c | <span style=color:#7f848e>#count the occurrence of each fragment length</span>
</span></span><span style=display:flex><span> 
</span></span><span style=display:flex><span>sort -b -k2,2n | <span style=color:#7f848e>#sort based on fragment length</span>
</span></span><span style=display:flex><span> 
</span></span><span style=display:flex><span>sed -e <span style=color:#98c379>&#39;s/^[ \t]*//&#39;</span> <span style=color:#7f848e>#remove leading blank of the standard output</span>
</span></span></code></pre></div><p>In the output, the 2nd column is the fragment length, and the 1st column is the counts. Open <code>fragment_length_count.txt</code> with Excel, and simply plot the 2nd column against the 1st column:</p><p><img src=/images/000/atac_isize_excel.png alt></p><p>Okay, I guess this is not really about ATAC-seq, it is more about command line usage, and I admit, it is not as beautiful as the plot generated by <strong><a href=https://matplotlib.org/>matplotlib</a></strong> or <strong><a href=https://www.r-project.org/>R</a></strong>.</p><p>This is kind of like a histogram with 1-bp bin, which sometimes could be very spiky without smoothing.</p><p>In addition, many beginners don&rsquo;t know what to expect from the library profiles (before sequencing) when running on a gel or Bioanalyzer/tapeStation etc. From our experience, we often see at least three peaks:</p><ol><li>The sharp peak around 200 bp, representing the open chromatin;</li><li>The peak around 350 bp, representing mono-nucleosome;</li><li>The peak around 550 bp, representing di-nucleosome;</li></ol><p>Sometimes, we also see more peaks larger than 550 bp (not always). It could be a combination of large tagmented fragments and the artefacts of the Bioanalyzer (it is not a normal log scale like agarose gel). Click <strong><a href=https://github.com/dbrg77/plate_scATAC-seq/tree/master/figures>here</a></strong> to see a few examples of Bioanalyzer profiles from successful ATAC-seq experiments. Check the pdf files with names <code>rep{1..10}_bioanlayzer.pdf</code>. Note those are libraries where the Illumina Nextera adapters (136 bp) have been added.</p></p></div><div class=prev-next></div></div></main><footer class=footer><span>&copy; 2025 Not A Rocket Scientist</span>
<span>Made with &#10084;&#65039; using <a target=_blank href=https://github.com/526avijitgupta/gokarna>Gokarna</a></span></footer></body></html>