<!DOCTYPE html>
<html lang="en"><head>
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/site.webmanifest">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>Understanding The Chromap Summary File</title>
    <meta name="description" content="Genomics | Bioinformatics | Epigenetics">
    <meta name="keywords" content='blog, gokarna, hugo, data science, biology, genomics, single cell genomics, sequencing library, illumina sequencing, chromap, ATAC-seq'>

    <meta property="og:url" content="https://notarocketscientist.xyz/posts/2024-07-23-understanding-the-chromap-summary-file/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="Understanding The Chromap Summary File">
    <meta property="og:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="og:image" content="/images/profile_avatar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="Understanding The Chromap Summary File">
    <meta name="twitter:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="twitter:domain" content="https://notarocketscientist.xyz/posts/2024-07-23-understanding-the-chromap-summary-file/">
    <meta property="twitter:url" content="https://notarocketscientist.xyz/posts/2024-07-23-understanding-the-chromap-summary-file/">
    <meta name="twitter:image" content="/images/profile_avatar.png">

    
    <link rel="canonical" href="https://notarocketscientist.xyz/posts/2024-07-23-understanding-the-chromap-summary-file/" />

    <link rel="stylesheet" type="text/css" href="/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="/css/dark.css">

    <script src="/js/svg-injector.min.js"></script>
    <script src="/js/feather-icons.min.js"></script>
    <script src="/js/main.js"></script>

    
    
        <!-- KaTeX -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css" integrity="sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ" crossorigin="anonymous">
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js" integrity="sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB" crossorigin="anonymous"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js" integrity="sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
    <script>
        document.addEventListener("DOMContentLoaded", function() {
            renderMathInElement(document.body, {
                delimiters: [
                    {left: "$$", right: "$$", display: true},
                    {left: "$", right: "$", display: false}
                ]
            });
        });
    </script>
  
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://notarocketscientist.xyz">
                <img src="https://notarocketscientist.xyz/images/profile_avatar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://notarocketscientist.xyz">Not A Rocket Scientist</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/about/"><span data-feather='user'></span> About </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/posts/"><span data-feather='list'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/projects/"><span data-feather='square'></span> Projects </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
            </div>
            
            <div class="nav-link">
                <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/about/"><span data-feather='user'></span> About </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/posts/"><span data-feather='list'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/projects/"><span data-feather='square'></span> Projects </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>Understanding The Chromap Summary File</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            July 23, 2024
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/single-cell-genomics">single cell genomics</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/sequencing-library">sequencing library</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/illumina-sequencing">illumina sequencing</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/chromap">chromap</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/atac-seq">ATAC-seq</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <h2 id="background">Background</h2>
<p>When talking about scATAC-seq data preprocessing, I guess the first method that most people choose is <a href="https://support.10xgenomics.com/single-cell-atac/software/overview/welcome"><strong>Cell Ranger ATAC</strong></a>. The 10x Genomics Single Cell ATAC kit is probably the most widely used commercial method for doing scATAC-seq, and <strong>Cell Ranger ATAC</strong> provides a very convenient way of converting the raw sequencing reads to some sort of count matrix for downstream in-depth analyses. The problems with <strong>Cell Ranger ATAC</strong> are that:</p>
<ul>
<li>It can only handle data from the 10x Genomics kit</li>
<li>It consumes too much resource and takes too long to run.</li>
</ul>
<p>Anybody that has used <strong>Cell Ranger ATAC</strong> should know what I&rsquo;m talking about. Then <a href="https://www.nature.com/articles/s41467-021-26865-w"><strong>Chromap</strong></a> was published. It is a light-weighted and really fast read aligner for single cell ATAC-seq and other chromatin profiling assays. We tried it when it was on <a href="https://www.biorxiv.org/content/10.1101/2021.06.18.448995v1">bioRxiv</a>, and we immediately started adapting our existing pipelines to use it for almost all chromatin assays. <strong>Chromap</strong> has an ATAC-seq mode, that is, the <code>--preset atac</code> option. If the ATAC mode is turned on, <strong>Chromap</strong> will perform all of the following steps automatically for us:</p>
<ul>
<li>Trim adapter sequences at the 3&rsquo; end of the reads</li>
<li>Set the maximum insert size between Read 1 &amp; Read 2 to be 2000 bp</li>
<li>Deduplicate the mapped read pairs at the single-cell level</li>
<li>Correct cell barcodes based on provided <code>whitelist</code></li>
<li>Shift read positions to make the start positions (5’) of the reads represent the actual Tn5 binding sites (not important for downstream peak calling, though)</li>
<li>Output read pairs (MAPQ &gt;= 30 by default) as the commonly-used <code>BED</code>-like file <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments"><strong>fragments.tsv</strong></a>. In addition, the <code>SAM</code> format is also supported.</li>
</ul>
<p>Then you can do whatever you want with the output files, which is really convenient and saves us from many preprocessing effort. It is much faster and more flexible than <strong>Cell Ranger ATAC</strong>. We are using <strong>Chromap</strong> to <a href="https://scg-lib-structs.readthedocs.io/en/latest/Epigenetics.html">do preprocessing for scATAC-seq data obtained from many different methods and platform</a>, <em>i.e.</em> not limited to the 10x scATAC-seq data.</p>
<p>However, what I do like about <strong>Cell Ranger ATAC</strong> is the per cell level QC information generated during the run, which is really helpful in terms of troubleshooting and getting a rough idea about our data quality. The file I&rsquo;m talking about is <code>singlecell.csv</code> generated during a run. Using <strong>Cell Ranger ATAC</strong> <code>v2.0.0</code> as the example, the file should be located at:</p>
<pre tabindex="0"><code>output_prefix/outs/singlecell.csv
</code></pre><p>Here are the top 10 lines of <code>singlecell.csv</code> from one of our samples:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>barcode,total,duplicate,chimeric,unmapped,lowmapq,mitochondrial,nonprimary,passed_filters,is__cell_barcode,excluded_reason,TSS_fragments,DNase_sensitive_region_fragments,enhancer_region_fragments,promoter_region_fragments,on_target_fragments,blacklist_region_fragments,peak_region_fragments,peak_region_cutsites
</span></span><span style="display:flex;"><span>NO_BARCODE,18682896,5076127,27002,1691108,1137896,20372,1638,10728753,0,0,0,0,0,0,0,0,0,0
</span></span><span style="display:flex;"><span>AAACGAAAGAAACGCC-1,8,1,0,0,2,0,0,5,0,0,2,0,0,0,2,0,3,6
</span></span><span style="display:flex;"><span>AAACGAAAGAAAGCAG-1,4,3,0,0,0,0,0,1,0,2,0,0,0,0,0,0,0,0
</span></span><span style="display:flex;"><span>AAACGAAAGAAAGGGT-1,42,11,0,1,1,3,0,26,0,3,4,0,0,0,4,0,19,36
</span></span><span style="display:flex;"><span>AAACGAAAGAAATCTG-1,5,3,0,0,0,0,0,2,0,2,1,0,0,0,1,0,0,0
</span></span><span style="display:flex;"><span>AAACGAAAGAAATGGG-1,3,1,0,0,0,0,0,2,0,0,1,0,0,0,1,0,1,2
</span></span><span style="display:flex;"><span>AAACGAAAGAACAGGA-1,2,1,0,0,0,0,0,1,0,2,1,0,0,0,1,0,0,0
</span></span><span style="display:flex;"><span>AAACGAAAGAACCCGA-1,4,3,0,0,0,0,0,1,0,2,0,0,0,0,0,0,0,0
</span></span><span style="display:flex;"><span>AAACGAAAGAACGCCA-1,2,0,0,0,0,0,0,2,0,2,0,0,0,0,0,0,0,0
</span></span></code></pre></div><p>You cannot easily get the information at this level of details from other programs. Last year, <strong>Chromap</strong> <code>v0.2.5 (r473)</code> added the <code>--summary</code> option that can provide similar information that we want. That is really helpful. Having quickly tested the new option, I realised that the numbers are a bit confusing and need some further explanation. For example, it is difficult to know if the numbers are from mutually exclusive categories in the summary. Here I documented how I read those numbers to help me understand the program a bit better. We are using the latest version <code>0.2.6-r490</code> at this time of writing.</p>
<p>First, when running <strong>Chromap</strong> interactively, we see that there are some basic information about the reads printed to the screen. Those things are the <strong>standard error</strong> during the run. Let’s use the single-cell ATAC data from <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-11264/sdrf">a mouse cortex</a> that we <a href="https://www.nature.com/articles/s41592-022-01601-4">published previously</a> as an example. Here we capture the standard error into a text file by <code>2&gt;</code>, and the command that we routinely use is like this:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>chromap -t <span style="color:#d19a66">20</span> -x ref.idx -r ref.fa <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        --preset atac --summary summary.csv <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        -1 mCortex_ATAC_S1_L001_R1_001.fastq.gz <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        -2 mCortex_ATAC_S1_L001_R2_001.fastq.gz <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        -b mCortex_ATAC_S1_L001_I2_001.fastq.gz <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        --barcode-whitelist 737K-cratac-v1_rc.txt <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        -o chromap_outs/fragments.tsv <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        1&gt; chromap.stdout 2&gt; chromap.stderr <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>        <span style="color:#56b6c2">&amp;&amp;</span> bgzip chromap_outs/fragments.tsv
</span></span></code></pre></div><p>There are two things we need to investigate:</p>
<ol>
<li>We see some common information about the reads in <code>chromap.stderr</code> and <code>summary.csv</code>. Are they consistent?</li>
<li>There are still other pieces of QC information missing in the <code>summary.csv</code> file, like mapping rate, sequence saturation (or duplications rate), mapping rate, and fraction of reads in peaks (FRiP).</li>
</ol>
<h2 id="consistency-between-stderr-and-summary">Consistency between stderr and summary</h2>
<p>The first problem is relatively easy to figure out. There are some inconsistencies between the quality metrics from <code>chromap.stderr</code> and <code>summary.csv</code>. The numbers in <code>chromap.stderr</code> represent the read number, but the numbers in <code>summary.csv</code> mean <strong>fragments</strong>, that is, properly mapped read pairs. Here are the lines related to the reads at the end of  <code>chromap.stderr</code> :</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 1</span><span>Number of reads: 951163184.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 2</span><span>Number of mapped reads: 810898762.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 3</span><span>Number of uniquely mapped reads: 792439756.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 4</span><span>Number of reads have multi-mappings: 18459006.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 5</span><span>Number of candidates: 16143953592.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 6</span><span>Number of mappings: 810898762.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 7</span><span>Number of uni-mappings: 792439756.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 8</span><span>Number of multi-mappings: 18459006.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 9</span><span>Number of barcodes in whitelist: 438281445.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">10</span><span>Number of corrected barcodes: 18669661.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">11</span><span>Sorted, deduped and outputed mappings in 268.81s.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">12</span><span># uni-mappings: 161581509, # multi-mappings: 6588743, total: 168170252.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">13</span><span>Number of output mappings (passed filters): 158101734</span></span></code></pre></div>
<p>The <code>summary.csv</code> file contains the per-cell level QC information, and here are the top 10 lines of the file:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>barcode,total,duplicate,unmapped,lowmapq
</span></span><span style="display:flex;"><span>CAAAAAAAAAAAAAAA,605,0,605,0
</span></span><span style="display:flex;"><span>AACCAAAAAAAAAAAA,77,0,77,0
</span></span><span style="display:flex;"><span>CACTAAAAACAAAAAG,1,0,1,0
</span></span><span style="display:flex;"><span>TGCAAAAAACAAAAAG,1,0,1,0
</span></span><span style="display:flex;"><span>ACACAAAAAGAAAACA,1,0,1,0
</span></span><span style="display:flex;"><span>GGTAAAAAACAAAAAG,1,0,1,0
</span></span><span style="display:flex;"><span>GAAAAAAAAAAAAAAA,830,0,830,0
</span></span><span style="display:flex;"><span>CTTAAAAAATAAAACG,1,0,1,0
</span></span><span style="display:flex;"><span>AAAAAAAACAAAAAGA,3,0,3,0
</span></span></code></pre></div><p>The information from the two files should be consistent. Now let’s figure out what those numbers mean in each file. If we sum up the reads in each column from  <code>summary.csv</code>, we can get:</p>
<table>
<thead>
<tr>
<th style="text-align:left">Column</th>
<th style="text-align:left">Sum</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><strong>total</strong></td>
<td style="text-align:left">475,581,592</td>
</tr>
<tr>
<td style="text-align:left"><strong>duplicate</strong></td>
<td style="text-align:left">230,290,114</td>
</tr>
<tr>
<td style="text-align:left"><strong>unmapped</strong></td>
<td style="text-align:left">70,132,211</td>
</tr>
<tr>
<td style="text-align:left"><strong>lowmapq</strong></td>
<td style="text-align:left">17,057,533</td>
</tr>
</tbody>
</table>
<p>Now let’s first check the number of reads in the original <code>fastq</code> files:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-console" data-lang="console"><span style="display:flex;"><span>$ zcat mCortex_ATAC_S1_L001_R1_001.fastq.gz | wc -l
</span></span><span style="display:flex;"><span>1902326368
</span></span></code></pre></div><p>It turns out that we have 475,581,592 (1,902,326,368/4) reads for each <code>R1</code>, <code>R2</code> and <code>I2</code>. This number is the same as the sum of the column <strong>total</strong>. Therefore, it is obvious that the numbers in the <code>summary.csv</code> file mean <strong>read pairs</strong> or <strong>fragments</strong>. The numbers in the <code>chromap.stderr</code> file mean <strong>reads</strong>. With that in mind, let’s use the numbers in <code>summary.csv</code> to reproduce the numbers in <code>chromap.stderr</code>.</p>
<p>The first line is:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>Number of reads: 951163184.
</span></span></code></pre></div><p>which is basically <strong>total</strong> * 2:</p>
<p>475,581,592 $\times$ 2 = 951,163,184</p>
<p>We can look at the 2nd, the 3rd and the 4th lines at the same time:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>Number of mapped reads: 810898762.
</span></span><span style="display:flex;"><span>Number of uniquely mapped reads: 792439756.
</span></span><span style="display:flex;"><span>Number of reads have multi-mappings: 18459006.
</span></span></code></pre></div><p>We noticed that &ldquo;Number of mapped reads&rdquo; is the sum of &ldquo;Number of uniquely mapped reads&rdquo; and &ldquo;Number of reads have multi-mappings&rdquo;:</p>
<p>810,898,762 = 792,439,756 + 18,459,006</p>
<p>The &ldquo;Number of mapped reads&rdquo; is consistent with the <code>summary.csv</code> file, which is (<strong>total</strong> - <strong>unmapped</strong>) * 2:</p>
<p>(475,581,592 - 70,132,211) $\times$ 2 = 810,898,762</p>
<p>Then the numbers of uniquely mapped and multi-mapping reads are 792,439,756 and 18,459,006, respectively. Since we know that <a href="https://lh3lh3.users.sourceforge.net/mapuniq.shtml">mapping uniqueness</a> is related to <strong>MAPQ</strong>, multi-mapping reads should have low <strong>MAPQ</strong>. Based on this logic, we should expect &ldquo;Number of uniquely mapped reads&rdquo; to be (<strong>total</strong> - <strong>unmapped</strong> - <strong>lowmapq</strong>) * 2 and &ldquo;Number of reads have multi-mappings&rdquo; to be <strong>lowmapq</strong> * 2. However, it turns out:</p>
<p>(475,581,592 - 70,132,211 - 17,057,533) $\times$ 2 = 776,783,696 &lt; 792,439,756</p>
<p>17,057,533 $\times$ 2 = 34,115,066 &gt; 18,459,006</p>
<p>Now we realised that <strong>Chromap</strong> has a <code>-q</code> flag to control the minimum <strong>MAPQ</strong> to send to the output. The default is <code>-q 30</code>, meaning any read with <strong>MAPQ</strong> &lt; 30 is considered to be <strong>lowmapq</strong>. However, I guess the definitions of &ldquo;uniquely mapped reads&rdquo; and &ldquo;multi-mapped reads&rdquo; are <strong>MAPQ</strong> &gt; 0 and <strong>MAPQ</strong> = 0, respectively. Under the current setting, we have some &ldquo;uniquely mapped reads&rdquo; in the <strong>lowmapq</strong> category.</p>
<p>Is this the case? We re-ran the program with the <code>-q 1</code> flag and redirect the standard error and summary to <code>q1_chromap.stderr</code> and <code>q1_summary.csv</code>, respectively. In this case, the <code>q1_chromap.stderr</code> remain almost unchanged, except the last line:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 1</span><span>Number of reads: 951163184.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 2</span><span>Number of mapped reads: 810898762.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 3</span><span>Number of uniquely mapped reads: 792439756.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 4</span><span>Number of reads have multi-mappings: 18459006.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 5</span><span>Number of candidates: 16143953592.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 6</span><span>Number of mappings: 810898762.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 7</span><span>Number of uni-mappings: 792439756.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 8</span><span>Number of multi-mappings: 18459006.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f"> 9</span><span>Number of barcodes in whitelist: 438281445.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">10</span><span>Number of corrected barcodes: 18669661.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">11</span><span>Sorted, deduped and outputed mappings in 268.81s.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">12</span><span># uni-mappings: 161581509, # multi-mappings: 6588743, total: 168170252.
</span></span><span style="display:flex;"><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#55595f">13</span><span>Number of output mappings (passed filters): 161219326</span></span></code></pre></div>
<p>which makes sense, since <code>-q</code> only changes the the filtering standards, <strong>NOT</strong> the behaviour of the program.</p>
<p>The column sums from the <code>q1_summary.csv</code> are:</p>
<table>
<thead>
<tr>
<th style="text-align:left">Column</th>
<th style="text-align:left">Sum</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><strong>total</strong></td>
<td style="text-align:left">475,581,592</td>
</tr>
<tr>
<td style="text-align:left"><strong>duplicate</strong></td>
<td style="text-align:left">234,736,148</td>
</tr>
<tr>
<td style="text-align:left"><strong>unmapped</strong></td>
<td style="text-align:left">70,132,211</td>
</tr>
<tr>
<td style="text-align:left"><strong>lowmapq</strong></td>
<td style="text-align:left">9,493,907</td>
</tr>
</tbody>
</table>
<p>Due to the <code>-q 1</code> flag, the <strong>lowmapq</strong> column is basically the number of reads with a <strong>MAPQ</strong> of 0. Now let&rsquo;s do a similar calculations of (<strong>total</strong> - <strong>unmapped</strong> - <strong>lowmapq</strong>) * 2 and <strong>lowmapq</strong> * 2, and we compared them to &ldquo;Number of uniquely mapped reads&rdquo; and &ldquo;Number of reads have multi-mappings&rdquo;, respectively:</p>
<p>(475,581,592 - 70,132,211 - 17,057,533) $\times$ 2 = 791,910,948 &lt; 792,439,756</p>
<p>94,93,907 $\times$ 2 = 18,987,814 &gt; 18,459,006</p>
<p>The numbers are still not the same, but they are close. There are 18,987,814 - 18,459,006 = 528,808 more reads in the <code>q1_summary.csv</code> file compared to <code>q1_chromap.stderr</code>, and 792,439,756 - 791,910,948 = 528,808 fewer reads in the <code>q1_summary.csv</code> file compared to <code>q1_chromap.csv</code>. It seems these 528,808 reads have <strong>MAPQ</strong> of 0s, but they are considered to be uniquely mapped reads. I asked the question in their GitHub repo (issue <a href="https://github.com/haowenz/chromap/issues/167">#167</a>). Thank <a href="https://github.com/mourisl">Li Song</a> for the explanation, now we understand that reads that are uniquely but poorly mapped to the genome are still considered as uniquely mapped but will be assigned a <strong>MAPQ</strong> of 0. That explains the discrepancies.</p>
<p>Now let&rsquo;s continue with <code>q1_chromap.stderr</code>. Lines 5-11 are straightforward, but we need to take extra care of of lines 9-10:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>Number of barcodes in whitelist: 438281445.
</span></span><span style="display:flex;"><span>Number of corrected barcodes: 18669661.
</span></span></code></pre></div><p>The sum of the total column from <code>q1_summary.csv</code> is 475,581,592, which is much higher than 438,281,445 + 18,669,661. Then we realised that there were 5,090,766 lines of record in the <code>q1_summary.csv</code> file:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ wc -l q1_summary.csv
</span></span><span style="display:flex;"><span><span style="color:#d19a66">5090766</span> q1_summary.csv
</span></span></code></pre></div><p>meaning there are 5,090,765 barcodes there, but we only have 737,280 barcodes in the whitelist <code>737K-cratac-v1_rc.txt</code>. It seems that barcoded not in the whitelist are also included in the summary. Let&rsquo;s figure out how many reads are there from barcodes in the whitelist and see if they are consistent with the <code>q1_chromap.stderr</code> file. This can be achieved by:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>join -t <span style="color:#98c379">&#34;,&#34;</span> -1 <span style="color:#d19a66">1</span> -2 <span style="color:#d19a66">1</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>    &lt;<span style="color:#56b6c2">(</span>sort 737K-cratac-v1_rc.txt<span style="color:#56b6c2">)</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>    &lt;<span style="color:#56b6c2">(</span>sed -n <span style="color:#98c379">&#39;2,$ p&#39;</span> q1_summary.csv | sort -t <span style="color:#98c379">&#34;,&#34;</span> -k1,1<span style="color:#56b6c2">)</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>    &gt; q1_summary_in_whitelist.csv
</span></span></code></pre></div><p>The sum of the column <strong>total</strong> from <code>q1_summary_in_whitelist.csv</code> is 456,955,085, which is so close but not the same to 438,281,445 + 18,669,661 = 456,951,106. It turns out that it is cause by <a href="https://github.com/haowenz/chromap/issues/167#issuecomment-2239465037">a known bug</a> of treating &lsquo;N&rsquo; in the barcode reads as &lsquo;A&rsquo;, and will be solved in the future.</p>
<p>The last two lines (lines 12 &amp; 13) are the numbers after deduplications, and they are the <strong>fragment</strong> numbers, which is a bit confusing, since all previous numbers are <strong>read</strong> numbers. Anyway, the last line says:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-plain" data-lang="plain"><span style="display:flex;"><span>Number of output mappings (passed filters): 161219326
</span></span></code></pre></div><p>which is consistent with the number of lines in the <code>fragment.tsv</code> file:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ wc -l fragments.tsv
</span></span><span style="display:flex;"><span><span style="color:#d19a66">161219326</span> fragments.tsv
</span></span></code></pre></div><p>With all the above information, I can finally put all numbers together in the diagram below from my original run with the default <code>-q 30</code> setting. Note the numbers indicate fragment numbers, or read pairs:</p>
<p><img src="/images/2024-07-23/reads_categorisation.png" alt=""></p>
<h2 id="other-qc-metrics">Other QC metrics</h2>
<p>The second problem is tha the <code>summary.csv</code> file does not contain all information that we want, such as the mapping rate, duplication rate, the number of unique fragments <em>etc.</em> Therefore we need to figure out by ourselves. I think the mapping rate, the duplication rate and the number of total unique fragments (nuclear + chrM) are straightforward to calculate. Before, we proceed, let&rsquo;s again do some sanity check.</p>
<p>To this end, let&rsquo;s use the cell <code>CTCGGATCTTACGTTG</code> as an example. First, we check its basic information in the <code>summary.csv</code> file:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ grep CTCGGATCTTACGTTG summary.csv
</span></span><span style="display:flex;"><span>CTCGGATCTTACGTTG,151143,87981,16097,4993
</span></span></code></pre></div><p>meaning it has a total of 151,143 fragments, of which 87,981 are duplicates, 16,097 are unmapped, and 4,993 have a MAPQ less than 30. What does it say in the <code>fragments.tsv</code> file? Let&rsquo;s check a few things.</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ grep CTCGGATCTTACGTTG fragments.tsv | addCols stdin
</span></span><span style="display:flex;"><span>0.00 2996956460198.00 2996966152396.00   0.00 130053.00
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>$ grep CTCGGATCTTACGTTG fragments.tsv | wc -l
</span></span><span style="display:flex;"><span><span style="color:#d19a66">42072</span>
</span></span></code></pre></div><p>The <code>addCols</code> program is a convenient binary from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/">UCSC utilities</a> that computes the sum of each column of a file or from standard input (stdin). In this case, the fifth column of the <code>fragments.tsv</code> file is the number of fragments at a particular position. The sum of them should be the total number of mapped fragments that passed filters in that cell, which is 130,053. This is equal to total - unmapped - lowmapq:</p>
<p>151,143 - 16,097 - 4,993 = 130,053</p>
<p>which is consistent with the summary. Good! The number of lines of the cell is 42,072, which should be the total number of fragments in that cell after deduplication. It is equal to total - unmapped - lowmapq - duplicate:</p>
<p>151,143 - 16,097 - 4,993 - 87,981 = 42,072</p>
<p>which is consistent with the summary. Again, good!</p>
<p>Now that we figure out the source of the numbers, we could compute some missing QC metrics. We can compute the following simply using the information from the <code>summary.csv</code> file:</p>
<p><strong>Overall alignment (mapping) rate:</strong> (total - unmapped)/total</p>
<p><strong>Total unique fragments:</strong> total - duplicate - unmapped - lowmapq</p>
<p><strong>Duplication rate (saturation):</strong> duplicate/(total - unmapped - lowmapq)</p>
<p>Of course, other QC metrics like chrM%, FRiP and TSS enrichment score are also important. They are slightly more difficult to get at this stage and I will write a new post to cover that<sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup>.</p>
<div class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1">
<p><strong>Update:</strong> Now that I think about it, I think the metrics like chrM%, FRiP and TSS enrichment score, are probably better calculated by dedicated downstream analysis tools, such as <a href="https://github.com/kaizhang/SnapATAC2">SnapATAC2</a> and <a href="https://github.com/stuart-lab/signac">Signac</a>.&#160;<a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</div>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>

<aside class="post-toc">
    <nav id="toc">
        <nav id="TableOfContents">
  <ul>
    <li><a href="#background">Background</a></li>
    <li><a href="#consistency-between-stderr-and-summary">Consistency between stderr and summary</a></li>
    <li><a href="#other-qc-metrics">Other QC metrics</a></li>
  </ul>
</nav>
    </nav>
</aside>



    

        </main><footer class="footer">
    <span>&copy; 2025 Not A Rocket Scientist</span>
    <span>
        Made with &#10084;&#65039; using <a target="_blank" href="https://github.com/526avijitgupta/gokarna">Gokarna</a>
    </span>
</footer>
</body>
</html>
