<!DOCTYPE html>
<html lang="en"><head>
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/site.webmanifest">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>The _alt Chromosomes In The Reference Genome</title>
    <meta name="description" content="Genomics | Bioinformatics | Epigenetics">
    <meta name="keywords" content='blog, gokarna, hugo, data science, biology, genomics, bioinformatics, preprocessing'>

    <meta property="og:url" content="https://dbrg77.github.io/posts/2024-01-05-the-alt-chromosomes-in-the-reference-genome/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="The _alt Chromosomes In The Reference Genome">
    <meta property="og:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="og:image" content="/images/profile_avatar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="The _alt Chromosomes In The Reference Genome">
    <meta name="twitter:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="twitter:domain" content="https://dbrg77.github.io/posts/2024-01-05-the-alt-chromosomes-in-the-reference-genome/">
    <meta property="twitter:url" content="https://dbrg77.github.io/posts/2024-01-05-the-alt-chromosomes-in-the-reference-genome/">
    <meta name="twitter:image" content="/images/profile_avatar.png">

    
    <link rel="canonical" href="https://dbrg77.github.io/posts/2024-01-05-the-alt-chromosomes-in-the-reference-genome/" />

    <link rel="stylesheet" type="text/css" href="/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="/css/dark.css">

    <script src="/js/svg-injector.min.js"></script>
    <script src="/js/feather-icons.min.js"></script>
    <script src="/js/main.js"></script>

    
    
        <!-- KaTeX -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css" integrity="sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ" crossorigin="anonymous">
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js" integrity="sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB" crossorigin="anonymous"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js" integrity="sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
    <script>
        document.addEventListener("DOMContentLoaded", function() {
            renderMathInElement(document.body, {
                delimiters: [
                    {left: "$$", right: "$$", display: true},
                    {left: "$", right: "$", display: false}
                ]
            });
        });
    </script>
  
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://dbrg77.github.io">
                <img src="https://dbrg77.github.io/images/profile_avatar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://dbrg77.github.io">Not A Rocket Scientist</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/about/"><span data-feather='user'></span> About </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/posts/"><span data-feather='list'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/projects/"><span data-feather='square'></span> Projects </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
            </div>
            
            <div class="nav-link">
                <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/about/"><span data-feather='user'></span> About </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/posts/"><span data-feather='list'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/projects/"><span data-feather='square'></span> Projects </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>The _alt Chromosomes In The Reference Genome</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            January 5, 2024
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://dbrg77.github.io/tags/bioinformatics">bioinformatics</a></li>
        
            <li class="post-tag"><a href="https://dbrg77.github.io/tags/preprocessing">preprocessing</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <p>Early this week, I was re-organising my bookmarks in the browser and I came across <a href="https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use"><strong>this post</strong></a> by Heng Li published more than 6 years ago. It reminded me one of the problem I faced ~10 years ago when I first started using the <code>GRCh38/hg38</code> version of the human reference genome for RNA-seq analysis. Let&rsquo;s have a look what the problem was.</p>
<p>Now if we go to the <a href="https://hgdownload.soe.ucsc.edu/downloads.html">UCSC genome browser</a> to download the reference genome data, we see that there are a couple of options:</p>
<p><img src="/images/2024-01-05/ucsc_human_ref_download.png" alt=""></p>
<p>Let&rsquo;s get the <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/">Analysis Set</a> for ease of use since we are doing NGS. By the way, I don&rsquo;t think this &ldquo;Analysis Set&rdquo; thing existed 10 years ago when I encountered the problem. In the download link, we see there are two versions:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-console" data-lang="console"><span style="display:flex;"><span>.
</span></span><span style="display:flex;"><span>├── hg38.analysisSet.chroms.tar.gz
</span></span><span style="display:flex;"><span>└── hg38.fullAnalysisSet.chroms.tar.gz
</span></span></code></pre></div><p>If we look at the content, we will realise that the following sequences are in both versions:</p>
<ul>
<li>Chromosomes
<ul>
<li>These are the sequences from regular reference chromosomes</li>
<li>They are <code>chr{1..22}.fa</code>, <code>chrX.fa</code>, <code>chrY.fa</code> and <code>chrM.fa</code></li>
<li>In addition, <code>chrEBV.fa</code> is also present as a &ldquo;decoy&rdquo; sequence to improve the accuracy of mapping</li>
</ul>
</li>
<li>Unlocalised scaffolds
<ul>
<li>These are the sequences that we know associated with a particular chromosome, but the exact location and orientation are unknown.</li>
<li>They are named in the format of <code>chr{chromosome number or name}_{sequence_accession}v{sequence_version}_random</code></li>
<li>There are a total of 42 of them, <em>e.g.</em> <code>chr1_KI270706v1_random.fa</code></li>
</ul>
</li>
<li>Unplaced scaffolds
<ul>
<li>These are the sequences are not associated with any chromosomes</li>
<li>They are named in the format of <code>chrUn_{sequence_accession}v{sequence_version}</code></li>
<li>There are a total of 127 of them, <em>e.g.</em> <code>chrUn_KI270302v1.fa</code></li>
</ul>
</li>
</ul>
<p>Those three types of sequences make up what we called the <strong>primary assembly</strong> version of a genome. We can think <code>hg38.analysisSet.chroms.tar.gz</code> as the primary assembly of the <code>hg38</code> version of the human genome. On top of those, there are some extra sequences from <code>hg38.fullAnalysisSet.chroms.tar.gz</code>, they are:</p>
<ul>
<li>Alternate loci scaffolds
<ul>
<li>These are the alternative haplotypes that we know are present in the population</li>
<li>They are extremely similar to loci in the primary assembly</li>
<li>They are named in the format of <code>chr{chromosome number or name}_{sequence_accession}v{sequence_version}_alt</code></li>
<li>There are a total of 261 of them, <em>e.g.</em> <code>chr15_KI270905v1_alt.fa</code></li>
</ul>
</li>
</ul>
<p>It is those <code>_alt</code> sequences that cause the problem. Let&rsquo;s use the sample <a href="https://www.ebi.ac.uk/ena/browser/view/ERR315391"><strong>ERR3153919</strong></a> from <a href="https://www.ebi.ac.uk/ena/browser/home">ENA</a> for the demonstration. It is the RNA-seq experiment from the H1 human embryonic stem cell (hESC) line. I&rsquo;m using <a href="https://github.com/alexdobin/STAR">STAR</a> <code>v2.7.11a</code> to quantify the gene expression.</p>
<p>To start with the process, we make new directories to hold the reference genome data and raw RNA-seq data (FASTQ files). From here, we simply call the <strong>fullAnalysisSet</strong> reference genome &ldquo;full set&rdquo;, and the <strong>analysisSet</strong> &ldquo;no alt&rdquo;. In addition, we also need the gene annotation file, and we choose to use the RefSeq annotation:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#7f848e"># make three directories to hold ref and actual data</span>
</span></span><span style="display:flex;"><span>mkdir full_set no_alt fastq gtf
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># download the reference genomes and raw data</span>
</span></span><span style="display:flex;"><span>wget -c <span style="color:#98c379">&#39;https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.analysisSet.chroms.tar.gz&#39;</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     -O no_alt/hg38.analysisSet.chroms.tar.gz
</span></span><span style="display:flex;"><span>wget -c <span style="color:#98c379">&#39;https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.fullAnalysisSet.chroms.tar.gz&#39;</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     -O full_set/hg38.fullAnalysisSet.chroms.tar.gz
</span></span><span style="display:flex;"><span>wget -c <span style="color:#98c379">&#39;https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ncbiRefSeq.gtf.gz&#39;</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     -O gtf/hg38.ncbiRefSeq.gtf.gz
</span></span><span style="display:flex;"><span>wget -c <span style="color:#98c379">&#39;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR315/ERR315391/ERR315391_1.fastq.gz&#39;</span>
</span></span><span style="display:flex;"><span>     -O fastq/ERR315391_1.fastq.gz
</span></span><span style="display:flex;"><span>wget -c <span style="color:#98c379">&#39;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR315/ERR315391/ERR315391_2.fastq.gz&#39;</span>
</span></span><span style="display:flex;"><span>     -O fastq/ERR315391_2.fastq.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># unzip the gtf file</span>
</span></span><span style="display:flex;"><span>gunzip gtf/hg38.ncbiRefSeq.gtf.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># extract reference fasta files</span>
</span></span><span style="display:flex;"><span>tar zxf no_alt/hg38.analysisSet.chroms.tar.gz -C no_alt/
</span></span><span style="display:flex;"><span>tar zxf full_set/hg38.fullAnalysisSet.chroms.tar.gz -C full_set/
</span></span></code></pre></div><p>After that, our current directory looks like this:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>.
</span></span><span style="display:flex;"><span>├── fastq
</span></span><span style="display:flex;"><span>│   ├── ERR3153919_1.fastq.gz
</span></span><span style="display:flex;"><span>│   └── ERR3153919_2.fastq.gz
</span></span><span style="display:flex;"><span>├── full_set
</span></span><span style="display:flex;"><span>│   ├── hg38.fullAnalysisSet.chroms
</span></span><span style="display:flex;"><span>│   │   ├── chr10.fa
</span></span><span style="display:flex;"><span>│   │   ├── ........
</span></span><span style="display:flex;"><span>│   │   └── chrY_KI270740v1_random.fa
</span></span><span style="display:flex;"><span>│   └── hg38.fullAnalysisSet.chroms.tar.gz
</span></span><span style="display:flex;"><span>├── gtf
</span></span><span style="display:flex;"><span>│   └── hg38.ncbiRefSeq.gtf
</span></span><span style="display:flex;"><span>└── no_alt
</span></span><span style="display:flex;"><span>    ├── hg38.analysisSet.chroms
</span></span><span style="display:flex;"><span>    │   ├── chr10.fa
</span></span><span style="display:flex;"><span>    │   ├── ........
</span></span><span style="display:flex;"><span>    │   └── chrY_KI270740v1_random.fa
</span></span><span style="display:flex;"><span>    └── hg38.analysisSet.chroms.tar.gz
</span></span></code></pre></div><p>Now, let&rsquo;s make star index:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#7f848e"># for full_set</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#d19a66">20</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --runMode genomeGenerate <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeDir full_set/star_index/ <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeFastaFiles full_set/hg38.fullAnalysisSet.chroms/*.fa <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --sjdbGTFfile gtf/hg38.ncbiRefSeq.gtf
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># for no_alt</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#d19a66">20</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --runMode genomeGenerate <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeDir no_alt/star_index/ <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeFastaFiles no_alt/hg38.analysisSet.chroms/*.fa <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --sjdbGTFfile gtf/hg38.ncbiRefSeq.gtf
</span></span></code></pre></div><p>Then we perform the actual alignment and quantification:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#7f848e"># for full_set</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#d19a66">20</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeDir full_set/star_index <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --readFilesIn fastq/ERR3153919_1.fastq.gz fastq/ERR3153919_2.fastq.gz <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --readFilesCommand zcat <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --quantMode GeneCounts <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --outFileNamePrefix ./star_outs/full_set/
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># for no_alt</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#d19a66">20</span> <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --genomeDir no_alt/star_index <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --readFilesIn fastq/ERR3153919_1.fastq.gz fastq/ERR3153919_2.fastq.gz <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --readFilesCommand zcat <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --quantMode GeneCounts <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>     --outFileNamePrefix ./star_outs/no_alt/
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span><span style="color:#7f848e"># index and sort the output sam file for future use</span>
</span></span><span style="display:flex;"><span>samtools view -u star_outs/full_set/Aligned.out.sam | <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>    samtools sort - -T tmp -o star_outs/full_set/Aligned.out.sorted.bam
</span></span><span style="display:flex;"><span>samtools index star_outs/full_set/Aligned.out.sorted.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>samtools view -u star_outs/no_alt/Aligned.out.sam | <span style="color:#98c379">\
</span></span></span><span style="display:flex;"><span><span style="color:#98c379"></span>    samtools sort - -T tmp -o star_outs/no_alt/Aligned.out.sorted.bam
</span></span><span style="display:flex;"><span>samtools index star_outs/no_alt/Aligned.out.sorted.bam
</span></span></code></pre></div><p>After the above processes are finished, we should have the following files in the <code>star_outs</code> directory:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>star_outs/
</span></span><span style="display:flex;"><span>├── full_set
</span></span><span style="display:flex;"><span>│   ├── Aligned.out.sam
</span></span><span style="display:flex;"><span>│   ├── Aligned.out.sorted.bam
</span></span><span style="display:flex;"><span>│   ├── Aligned.out.sorted.bam.bai
</span></span><span style="display:flex;"><span>│   ├── Log.final.out
</span></span><span style="display:flex;"><span>│   ├── Log.out
</span></span><span style="display:flex;"><span>│   ├── Log.progress.out
</span></span><span style="display:flex;"><span>│   ├── ReadsPerGene.out.tab
</span></span><span style="display:flex;"><span>│   └── SJ.out.tab
</span></span><span style="display:flex;"><span>└── no_alt
</span></span><span style="display:flex;"><span>    ├── Aligned.out.sam
</span></span><span style="display:flex;"><span>    ├── Aligned.out.sorted.bam
</span></span><span style="display:flex;"><span>    ├── Aligned.out.sorted.bam.bai
</span></span><span style="display:flex;"><span>    ├── Log.final.out
</span></span><span style="display:flex;"><span>    ├── Log.out
</span></span><span style="display:flex;"><span>    ├── Log.progress.out
</span></span><span style="display:flex;"><span>    ├── ReadsPerGene.out.tab
</span></span><span style="display:flex;"><span>    └── SJ.out.tab
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#d19a66">2</span> directories, <span style="color:#d19a66">16</span> files
</span></span></code></pre></div><p>Now we could create a scatter plot using the 2nd column of <code>ReadsPerGene.out.tab</code>, which is a tab-delimited file containing the read count of each gene in the sample. Plotting <strong>full_set</strong> vs <strong>no_alt</strong> in log scale, we have:</p>
<figure><img src="/images/2024-01-05/scatter.png" width="500px"/>
</figure>

<p>Well &hellip; If we look at the scatter, the vast majority of the genes align perfectly on the line $y=x$, which is good. However, 1761 genes are not on the line, most of which have orders of magnitudes more reads in the <strong>&ldquo;no alt&rdquo;</strong> version compared to the <strong>&ldquo;full set&rdquo;</strong>. The three genes at the outermost that are highlighted are <em>OCT4</em> (<em>POU5F1</em>), beta tubulin (<em>TUBB</em>) and 14-3-3$\varepsilon$ (<em>YWHAE</em>). Let&rsquo;s use the famous <em>POU5F1</em> gene as an example. If you work on embryonic stem cells or embryo development, you definitely know this gene. There were 23,992 reads mapped to this gene when using the &ldquo;no alt&rdquo; version, but only 851 reads were mapped to this gene when using the &ldquo;full set&rdquo; version of the genome. For the exact the same data set, the difference is huge. What is going on here? First, we could have a look at where the gene <em>POU5F1</em> is located:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>grep -w <span style="color:#98c379">&#39;gene_name &#34;POU5F1&#34;&#39;</span> gtf/hg38.ncbiRefSeq.gtf
</span></span></code></pre></div><p>Then we realise that there are multiple records of <em>POU5F1</em>:</p>
<pre tabindex="0"><code>chr6	ncbiRefSeq.2022-10-28	transcript	31164337	31167170	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_7&#34;;  gene_name &#34;POU5F1&#34;;
chr6	ncbiRefSeq.2022-10-28	transcript	31164337	31167170	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_7&#34;;  gene_name &#34;POU5F1&#34;;
chr6	ncbiRefSeq.2022-10-28	transcript	31164337	31167170	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_7&#34;;  gene_name &#34;POU5F1&#34;;
chr6	ncbiRefSeq.2022-10-28	transcript	31164337	31170682	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_7&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000256v2_alt	ncbiRefSeq.2022-10-28	transcript	2467753	2470601	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000256v2_alt	ncbiRefSeq.2022-10-28	transcript	2467753	2470601	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000256v2_alt	ncbiRefSeq.2022-10-28	transcript	2467753	2470601	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000256v2_alt	ncbiRefSeq.2022-10-28	transcript	2467753	2474114	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000255v2_alt	ncbiRefSeq.2022-10-28	transcript	2422370	2425218	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_2&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000255v2_alt	ncbiRefSeq.2022-10-28	transcript	2422370	2425218	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_2&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000255v2_alt	ncbiRefSeq.2022-10-28	transcript	2422370	2425218	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_2&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000255v2_alt	ncbiRefSeq.2022-10-28	transcript	2422370	2428732	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_2&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000254v2_alt	ncbiRefSeq.2022-10-28	transcript	2508452	2511300	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_3&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000254v2_alt	ncbiRefSeq.2022-10-28	transcript	2508452	2511300	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_3&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000254v2_alt	ncbiRefSeq.2022-10-28	transcript	2508452	2511300	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_3&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000254v2_alt	ncbiRefSeq.2022-10-28	transcript	2508452	2514816	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_3&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000253v2_alt	ncbiRefSeq.2022-10-28	transcript	2474838	2477686	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_4&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000253v2_alt	ncbiRefSeq.2022-10-28	transcript	2474838	2477686	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_4&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000253v2_alt	ncbiRefSeq.2022-10-28	transcript	2474838	2477686	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_4&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000253v2_alt	ncbiRefSeq.2022-10-28	transcript	2474838	2481202	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_4&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000252v2_alt	ncbiRefSeq.2022-10-28	transcript	2423653	2426501	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_5&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000252v2_alt	ncbiRefSeq.2022-10-28	transcript	2423653	2426501	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_5&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000252v2_alt	ncbiRefSeq.2022-10-28	transcript	2423653	2426501	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_5&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000252v2_alt	ncbiRefSeq.2022-10-28	transcript	2423653	2430016	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_5&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000251v2_alt	ncbiRefSeq.2022-10-28	transcript	2646761	2649609	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001285986.2_6&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000251v2_alt	ncbiRefSeq.2022-10-28	transcript	2646761	2649609	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_203289.6_6&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000251v2_alt	ncbiRefSeq.2022-10-28	transcript	2646761	2649609	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_001173531.3_6&#34;;  gene_name &#34;POU5F1&#34;;
chr6_GL000251v2_alt	ncbiRefSeq.2022-10-28	transcript	2646761	2653124	.	-	.	gene_id &#34;POU5F1&#34;; transcript_id &#34;NM_002701.6_6&#34;;  gene_name &#34;POU5F1&#34;;
</code></pre><p>As we can see, there are multiple transcripts (isoforms) for the same gene <em>POU5F1</em>, but the true record is those from <strong>chr6</strong>. Other records from <code>chr6_GL00025{1..6}v2_alt</code> are just alternative haplotypes in the population. They have nearly identical sequences to the segment from chr6. Let&rsquo;s see some reads mapped to <em>POU5F1</em>, starting from the &ldquo;no alt&rdquo; version:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ samtools view star_outs/no_alt/Aligned.out.sorted.bam chr6:31,164,337-31,167,170 | head -5
</span></span><span style="display:flex;"><span>ERR3153919.15777041	163	chr6	31062785	255	<span style="color:#e06c75">14M217932N60M1S</span>	<span style="color:#56b6c2">=</span>	31280754	218041	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.8228362	163	chr6	31062786	255	<span style="color:#e06c75">13M217932N60M2S</span>	<span style="color:#56b6c2">=</span>	31280842	218124	GCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCAT	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:111	nM:i:9
</span></span><span style="display:flex;"><span>ERR3153919.8525639	419	chr6	31062789	1	<span style="color:#e06c75">10M217932N65M</span>	<span style="color:#56b6c2">=</span>	31280833	218116	GGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCAT	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:3	HI:i:3	AS:i:115	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.13201791	419	chr6	31062789	1	<span style="color:#e06c75">10M217932N65M</span>	<span style="color:#56b6c2">=</span>	31280837	218116	GGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCAT	BB/B/&lt;BFFF/BB//FFFF&lt;////B///<span style="color:#98c379">&lt;&lt;F&lt;/F</span>&lt;FFB&lt;FB/FF&lt;//&lt;//B&lt;&lt;////&lt;B///&lt;FFBFBB/&lt;77&lt;F	NH:i:3	HI:i:3	AS:i:111	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.18818531	419	chr6	31062789	1	<span style="color:#e06c75">10M217932N65M</span>	<span style="color:#56b6c2">=</span>	31280836	218116	GGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCAT	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:3	HI:i:3	AS:i:112
</span></span></code></pre></div><p>Note the first read <code>ERR3153919.15777041</code> is a good one that has a mapping quality (MAPQ) of 255, indicating it is a uniquely mapped read. Let&rsquo;s look at where this read would map to in the &ldquo;full set&rdquo; version of the genome:</p>
<div class="highlight"><pre tabindex="0" style="color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>$ samtools view star_outs/full_set/Aligned.out.sorted.bam | grep -w <span style="color:#98c379">&#39;ERR3153919.15777041&#39;</span>
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6	31062785	0	<span style="color:#e06c75">14M217932N60M1S</span>	<span style="color:#56b6c2">=</span>	31280754	218041	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:7	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6	31280754	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	31062785	-218041	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:7	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6_GL000251v2_alt	2545163	0	<span style="color:#e06c75">14M216464N60M1S</span>	<span style="color:#56b6c2">=</span>	2761664	216573	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:4	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6_GL000251v2_alt	2761664	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2545163	-216573	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:4	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6_GL000252v2_alt	2322157	0	<span style="color:#e06c75">14M216300N60M1S</span>	<span style="color:#56b6c2">=</span>	2538494	216409	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:5	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6_GL000252v2_alt	2538494	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2322157	-216409	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:5	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6_GL000253v2_alt	2373292	0	<span style="color:#e06c75">14M216426N60M1S</span>	<span style="color:#56b6c2">=</span>	2589755	216535	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:6	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6_GL000253v2_alt	2589755	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2373292	-216535	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:6	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6_GL000254v2_alt	2406901	0	<span style="color:#e06c75">14M216511N60M1S</span>	<span style="color:#56b6c2">=</span>	2623449	216620	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:1	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6_GL000254v2_alt	2623449	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2406901	-216620	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:1	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	163	chr6_GL000255v2_alt	2320749	0	<span style="color:#e06c75">14M215351N60M1S</span>	<span style="color:#56b6c2">=</span>	2536137	215460	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:2	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	83	chr6_GL000255v2_alt	2536137	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2320749	-215460	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:2	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	419	chr6_GL000256v2_alt	2366233	0	<span style="color:#e06c75">14M216393N60M1S</span>	<span style="color:#56b6c2">=</span>	2582663	216502	GGCAGGATGACACAGGCAAGAAGCAGCTGGAGAAGGACTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCA	BBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:7	HI:i:3	AS:i:114	nM:i:10
</span></span><span style="display:flex;"><span>ERR3153919.15777041	339	chr6_GL000256v2_alt	2582663	0	<span style="color:#e06c75">72M3S</span>	<span style="color:#56b6c2">=</span>	2366233	-216502	CTTCAACAGCATGAAGAAGTACTGCCAGGTCATCCGCATCATTGCCCACACTCAGATGCGTCTGCTTCCTCTACG	BFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB	NH:i:7	HI:i:3	AS:i:114	nM:i:10
</span></span></code></pre></div><p>As we can see, the same read was mapped to 7 locations<sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup> in the &ldquo;full set&rdquo; version of the genome. As a result, the mapping quality of this read is 0. One of those alignments was set as the primary alignment, which is the one mapped to <code>chr6_GL000255v2_alt</code>. That is, the pair of reads whose flags are 163 and 83. I think the primary alignment is chosen randomly when all alignments are equally good, but I&rsquo;m not entirely sure. The rest of the 6 alignments all have the <code>0x100</code> flag, that is, <strong>not primary alignment</strong>. Because of this, the read <code>ERR3153919.15777041</code> will not be used for quantification. There are many reads like this.</p>
<p>If the &ldquo;full set&rdquo; version is used, we would think there might be something wrong with our experiments, because <em>POU5F1</em> should be highly expressed in any embryonic stem cells. This is actually a good thing so you can investigate what&rsquo;s going on. However, if we are dealing with some other samples, we may not notice the problem. I&rsquo;m not sure if the difference of those 1761 genes would affect our major conclusions. The dangerous situation is always that <strong>we don&rsquo;t know we don&rsquo;t know!</strong></p>
<p>Anyway, this is first reminder in the new year. I <a href="https://twitter.com/XiChenUoM/status/1742788444219187477">tweeted</a> about this, and you can think of this post as a more detailed version of the tweet with code.</p>
<p>P.S. When I finished writing this post, I was wondering why I did not encounter the problem when using <code>hg19</code> where the <code>_alt</code> chromosomes were also included in the reference. Then I realised I was using the <a href="https://support.illumina.com/sequencing/sequencing_software/igenome.html">Illumina iGenomes</a> package where those things are removed. Oh well &hellip; I guess the take-home message is know your package and read the <strong>README</strong> carefully.</p>
<div class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1">
<p>Note the data is in pair end format, so Read 1 and Read 2 have the same name. We have a total of 14 records, so it is 14/2 = 7 locations. We can also see the optional field <code>NH:i:7</code> indicated the read is mapped to 7 locations in the genome.&#160;<a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</div>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2025 Not A Rocket Scientist</span>
    <span>
        Made with &#10084;&#65039; using <a target="_blank" href="https://github.com/526avijitgupta/gokarna">Gokarna</a>
    </span>
</footer>
</body>
</html>
