<!DOCTYPE html>
<html lang="en"><head>
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/site.webmanifest">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>About FastQ Files Submission</title>
    <meta name="description" content="Genomics | Bioinformatics | Epigenetics">
    <meta name="keywords" content='blog, gokarna, hugo, data science, biology, genomics, bioinformatics, fastq, illumina-sequencing, sequencing library, single cell genomics'>

    <meta property="og:url" content="https://notarocketscientist.xyz/posts/2022-02-27-about-fastq-files-submission/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="About FastQ Files Submission">
    <meta property="og:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="og:image" content="/images/profile_avatar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="About FastQ Files Submission">
    <meta name="twitter:description" content="Genomics | Bioinformatics | Epigenetics">
    <meta property="twitter:domain" content="https://notarocketscientist.xyz/posts/2022-02-27-about-fastq-files-submission/">
    <meta property="twitter:url" content="https://notarocketscientist.xyz/posts/2022-02-27-about-fastq-files-submission/">
    <meta name="twitter:image" content="/images/profile_avatar.png">

    
    <link rel="canonical" href="https://notarocketscientist.xyz/posts/2022-02-27-about-fastq-files-submission/" />

    <link rel="stylesheet" type="text/css" href="/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="/css/dark.css">

    <script src="/js/svg-injector.min.js"></script>
    <script src="/js/feather-icons.min.js"></script>
    <script src="/js/main.js"></script>

    
    
        <!-- KaTeX -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css" integrity="sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ" crossorigin="anonymous">
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js" integrity="sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB" crossorigin="anonymous"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js" integrity="sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
    <script>
        document.addEventListener("DOMContentLoaded", function() {
            renderMathInElement(document.body, {
                delimiters: [
                    {left: "$$", right: "$$", display: true},
                    {left: "$", right: "$", display: false}
                ]
            });
        });
    </script>
  
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://notarocketscientist.xyz">
                <img src="https://notarocketscientist.xyz/images/profile_avatar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://notarocketscientist.xyz">Not A Rocket Scientist</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/about/"><span data-feather='user'></span> About </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/posts/"><span data-feather='list'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/projects/"><span data-feather='square'></span> Projects </a>
            </div>
            
            <div class="nav-link">
                <a href="https://notarocketscientist.xyz/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
            </div>
            
            <div class="nav-link">
                <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/about/"><span data-feather='user'></span> About </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/posts/"><span data-feather='list'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/projects/"><span data-feather='square'></span> Projects </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://notarocketscientist.xyz/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIO210"><span data-feather='bar-chart-2'></span>  </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://github.com/Teichlab/scg_lib_structs"><span data-feather='smile'></span>  </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>About FastQ Files Submission</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            February 27, 2022
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/bioinformatics">bioinformatics</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/fastq">fastq</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/illumina-sequencing">illumina-sequencing</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/sequencing-library">sequencing library</a></li>
        
            <li class="post-tag"><a href="https://notarocketscientist.xyz/tags/single-cell-genomics">single cell genomics</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <p>This post collects my thoughts about <code>fastq</code> files accumlated over the years, specifically focusing on what should be included during a submission. It is not about the format of fastq files. If you are not sure about the format, you can check the <a href="https://en.wikipedia.org/wiki/FASTQ_format">Wikipedia page</a> and this excellent paper by <strong><a href="https://doi.org/10.1093/nar/gkp1137">Cock <em>et al.</em></a></strong> that contains very deteiled information and the history about the fastq file.</p>
<p>Let&rsquo;s first look at what you will get from a typical Illumina sequencing experiment. The library fragments you get is like this:</p>
<p><img src="/images/2022-02-27/lib_struct.pdf" alt=""></p>
<p>The middle dark thin lines are the DNA you want to sequence. The coloured boxes represent known sequences, often called adaptors, designed by Illumina. <strong><a href="https://teichlab.github.io/scg_lib_structs/methods_html/Illumina.html">Click here</a></strong> to look at the actual sequences if you are new to Illumina sequencing. The indices (<code>i5</code> and <code>i7</code>) shown above are short unique DNA sequences, often called barcodes, that will discrimate different samples. Those allow mixing and sequencing libraries from different samples at the same time. Now let&rsquo;s start the sequencing with a typical dual index pair end mode:</p>
<p><strong>Step 1</strong> is to denature the DNA, use the bottom strand as the template and add the <code>Read 1 primer</code> to start sequencing your DNA in the middle. This step yields a file called <code>Read1.fastq.gz</code> (note: striclty speaking, this is not the case, see the later section for the explanation):</p>
<p><img src="/images/2022-02-27/read1.pdf" alt=""></p>
<p><strong>Step 2</strong> is to use the bottom strand as the template again, add the <code>Index 1 primer</code> that is reverse complementary to the <code>Read 2 primer</code> and sequence the <code>i7</code> index. This step yields a file called <code>I1.fastq.gz</code>:</p>
<p><img src="/images/2022-02-27/index1.pdf" alt=""></p>
<p><strong>Step 3</strong> is to perform the cluster regeneration, use the top strand as the template, add the <code>Index 2 primer</code> that is reverse complementary to the <code>Read 1 primer</code> and sequence the <code>i5</code> index. This step yields a file called <code>I2.fastq.gz</code>:</p>
<p><img src="/images/2022-02-27/index2.pdf" alt=""></p>
<p><strong>Step 4</strong> is to use the top strand as the template again and add the <code>Read 2 primer</code> to sequence the other end of your DNA. This step yields a file called <code>Read2.fastq.gz</code>.</p>
<p><img src="/images/2022-02-27/read2.pdf" alt=""></p>
<p>The real situation is bit more complicated. You often only have one index (<code>i7</code>), so <strong>Step 3</strong> is omitted. Sometimes, you only do single end sequencing, so <strong>Step 4</strong> is omitted. In addition, the bases <code>A, C, G, T</code> are fluorescent signals. Those signals (images) are converted into DNA bases by a process called <code>base calling</code>, which is automatically done on the Illumina machine. The output you get directly from the sequencing machine is the result of the <code>base calling</code> step. The files are in the <code>BCL</code> (Binary Base Calling) format. To obtain the fastq files, you need to run the <a href="https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html">bcl2fatq</a> program provided by Illumina to covnert the <code>bcl</code> file to the fastq format.</p>
<p>If you haven&rsquo;t done that before, your sequencing core facility is probably doing those steps for you. They probably also seperate different samples for you based on the index you provide, a process called <code>demultiplexing</code>. Eventually, you get <code>Read1.fastq.gz</code> and <code>Read2.fastq.gz</code> per sample. Those are the actual DNA sequences you care about, and those files contain the useful information about your experiments. If you are running <code>bcl2fastq</code> by yourself, you may only get two, not four, fastq files per sample that are <code>Read1.fastq.gz</code> and <code>Read2.fastq.gz</code>. This is due to the default behaviour of the <code>bcl2fastq</code> program. To get all four fastq files, add the <code>--create-fastq-for-index-reads</code> flag when running <code>bcl2fastq</code>.</p>
<p>In the past, <code>Read1.fastq.gz</code> and <code>Read2.fastq.gz</code> were all we need, because all the actual information of the experiments are there. The index files (<code>I1.fastq.gz</code> and <code>I2.fastq.gz</code>) are not really useful once the demultiplexing step is done. That&rsquo;s why people just submit <code>Read1.fastq.gz</code> and <code>Read2.fastq.gz</code> to a public repostory, such as <a href="https://www.ebi.ac.uk/arrayexpress">ArrayExpress</a> and <a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a>, when they publish their papers. However, with the ever-growing complexity of current genomic methods, especially those in the field of single cell genomics, having just two fastq files is not enough.</p>
<p>In single cell genomics, the words <code>sample</code>, <code>index</code> and <code>barcode</code> are quite vague. Therefore, we now talk about the <code>cell barcode</code>, a short DNA sequence that is unique to a cell. All reads that are associated with a particular cell barcode are from the same cell. In any single cell methods, the <code>UMI</code>, if any, are usually in either <code>Read1.fastq.gz</code> or <code>Read2.fastq.gz</code>. However, the <code>cell barcode</code> can be anywhere in <code>Read1.fastq.gz</code>, <code>Read2.fastq.gz</code>, <code>I1.fastq.gz</code> and <code>I2.fastq.gz</code>. Sometime, a cell barcode is defined by the combination of sequences from those fastq files, this is especially true for all the &ldquo;split-pool&rdquo; based methods. To deal with this, I often put <code>N</code> in the <code>SampleSheet.csv</code>, a file that is used during the <code>bcl2fastq</code> step, like this:</p>
<pre tabindex="0"><code># this is an example from NextSeq
Sample_ID,Sample_Name,Sample_Plate,Sample_Well,Index_Plate,Index_Plate_Well,I7_Index_ID,index,I5_Index_ID,index2,Sample_Project,Description
foo,,,,,,N001,NNNNNNNN,S501,NNNNNNNN,,
</code></pre><p>If you do that, <code>bcl2fastq</code> will look for literal <code>N</code> in your index due to a known bug of the program, and anything matching <code>NNNNNNNN+NNNNNNNN</code> will be written into the sample <code>foo</code>. Hopefully, you will have very few or no reads in <code>foo</code>. All the useful reads will be written into the <code>Undetermined</code> sample. You will get the following four files per run:</p>
<pre tabindex="0"><code>Undetermined_R1.fastq.gz
Undetermined_R2.fastq.gz
Undetermined_I1.fastq.gz
Undetermined_I2.fastq.gz
</code></pre><p>All the information is there, and you can do whatever you need from this point. For me, I like the flexible settings of <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARsolo</a>, <a href="https://www.kallistobus.tools">kallisto|bustools</a> and <a href="https://github.com/sdparekh/zUMIs">zUMIs</a> <em>etc.</em> which allow preprocessing any methods in theory. For example, I don&rsquo;t really need to specify if the experiment protocol is using <code>10xv123</code>, <code>Drop-seq</code> or <code>CEL-seq</code> <em>etc.</em> I can use the combination of <code>soloCBstart</code>, <code>soloCBlen</code>, <code>soloUMIstart</code> and <code>soloUMIlen</code> in STARsolo to specify the locations of CB and UMI, even for future new methods. If you can&rsquo;t figure out the CB and UMI locations in a method, especially in a complicated single cell method, then can I shameless advertise my <a href="https://github.com/Teichlab/scg_lib_structs">scg_lib_structs</a> repository to you 😉? Okay, I guess my real point is we need to submit all four fastq files to public repositories.</p>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2025 Not A Rocket Scientist</span>
    <span>
        Made with &#10084;&#65039; using <a target="_blank" href="https://github.com/526avijitgupta/gokarna">Gokarna</a>
    </span>
</footer>
</body>
</html>
