<!doctype html><html lang=en><head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><link rel=apple-touch-icon sizes=180x180 href=/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/favicon-16x16.png><link rel=manifest href=/site.webmanifest><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><style>:root{--accent-color:#FF4D4D}</style><title>Getting Index Reads From SRA</title><meta name=description content="Genomics | Bioinformatics | Epigenetics"><meta name=keywords content='blog,gokarna,hugo,data science,biology,genomics,bioinformatics,fastq,illumina-sequencing,sratools,sequencing library,single cell genomics'><meta property="og:url" content="http://localhost:1313/posts/2022-07-26-getting-index-reads-from-sra/"><meta property="og:type" content="website"><meta property="og:title" content="Getting Index Reads From SRA"><meta property="og:description" content="Genomics | Bioinformatics | Epigenetics"><meta property="og:image" content="/images/profile_avatar.png"><meta name=twitter:card content="summary_large_image"><meta name=twitter:title content="Getting Index Reads From SRA"><meta name=twitter:description content="Genomics | Bioinformatics | Epigenetics"><meta property="twitter:domain" content="http://localhost:1313/posts/2022-07-26-getting-index-reads-from-sra/"><meta property="twitter:url" content="http://localhost:1313/posts/2022-07-26-getting-index-reads-from-sra/"><meta name=twitter:image content="/images/profile_avatar.png"><link rel=canonical href=http://localhost:1313/posts/2022-07-26-getting-index-reads-from-sra/><link rel=stylesheet type=text/css href=/css/normalize.min.css media=print onload='this.media="all"'><link rel=stylesheet type=text/css href=/css/main.css><link disabled id=dark-theme rel=stylesheet href=/css/dark.css><script src=/js/svg-injector.min.js></script><script src=/js/feather-icons.min.js></script><script src=/js/main.js></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css integrity=sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ crossorigin=anonymous><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js integrity=sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB crossorigin=anonymous></script><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js integrity=sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl crossorigin=anonymous onload=renderMathInElement(document.body)></script><script>document.addEventListener("DOMContentLoaded",function(){renderMathInElement(document.body,{delimiters:[{left:"$$",right:"$$",display:!0},{left:"$",right:"$",display:!1}]})})</script></head><body><script type=text/javascript>setThemeByUserPref()</script><header class=header><nav class=header-nav><div class=avatar><a href=http://localhost:1313/><img src=http://localhost:1313//images/profile_avatar.png alt=avatar></a></div><div class=nav-title><a class=nav-brand href=http://localhost:1313/>Not A Rocket Scientist</a></div><div class=nav-links><div class=nav-link><a href=http://localhost:1313/><span data-feather=home></span> Home</a></div><div class=nav-link><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></div><div class=nav-link><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></div><div class=nav-link><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></div><div class=nav-link><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></div><div class=nav-link><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></div><div class=nav-link><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></div><span class=nav-icons-divider></span><div class="nav-link dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only></span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></div><div class=nav-link id=hamburger-menu-toggle><span id=hamburger-menu-toggle-screen-reader-target class=sr-only>menu</span>
<a><span data-feather=menu></span></a></div><ul class="nav-hamburger-list visibility-hidden"><li class=nav-item><a href=http://localhost:1313/><span data-feather=home></span> Home</a></li><li class=nav-item><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></li><li class=nav-item><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></li><li class=nav-item><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></li><li class=nav-item><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></li><li class=nav-item><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></li><li class=nav-item><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></li><li class="nav-item dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only>theme</span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></li></ul></div></nav></header><main id=content><div class="post container"><div class=post-header-section><h1>Getting Index Reads From SRA</h1><small role=doc-subtitle></small><p class=post-date>July 26, 2022</p><ul class=post-tags><li class=post-tag><a href=http://localhost:1313/tags/bioinformatics>bioinformatics</a></li><li class=post-tag><a href=http://localhost:1313/tags/fastq>fastq</a></li><li class=post-tag><a href=http://localhost:1313/tags/illumina-sequencing>illumina-sequencing</a></li><li class=post-tag><a href=http://localhost:1313/tags/sratools>sratools</a></li><li class=post-tag><a href=http://localhost:1313/tags/sequencing-library>sequencing library</a></li><li class=post-tag><a href=http://localhost:1313/tags/single-cell-genomics>single cell genomics</a></li></ul></div><div class=post-content><p><p>A few month ago, I ranted about the <a href=/posts/2022-02-27-about-fastq-files-submission/><strong>FastQ files submission</strong></a> for data produced by complex genomic protocols. The main reason of the rant is that we tried to get some single cell data from <a href=https://www.ncbi.nlm.nih.gov/sra><strong>SRA</strong></a>, and we really needed the cell barcodes for the data preprocessing and analysis. However, the cell barcodes are stored in the indexing reads where it <em>seems</em> impossible to get from the <code>.sra</code> file. Recently, I finally found a way of getting those information &mldr; well &mldr; at least for some studies. Here, I&rsquo;m just documenting what happened.</p><p>First, the raw sequencing data is usually stored in the <code>sra</code> format, and you normally need the <code>fastq</code> files to start any analysis. To this end, you need the <code>fastq-dump</code> or <code>fasterq-dump</code> utilities from the <a href=https://github.com/ncbi/sra-tools><strong>sratools</strong></a>, which is not really well documented and contains many useful tricks that are obscure to non-routine users. For many studies, the sequencing is done using a pair-end mode, and you want to get both reads (<code>R1</code> and <code>R2</code>) in the fastq format. Let&rsquo;s start with using <code>fasterq-dump</code> as it is a more recent and, as the name suggested, a faster version of the previously used <code>fastq-dump</code> program. If you look at the help message of <code>fasterq-dump</code> you see two options for the purpose of getting paired fastq files:</p><pre tabindex=0><code>-S|--split-files      write reads into different files
-3|--split-3          writes single reads into special file
</code></pre><p>This is not very helpful. Luckily, I have used the old <code>fastq-dump</code> in the old days and seem to remember it has a better description. This is the help message from <code>fastq-dump</code>:</p><pre tabindex=0><code>--split-files           Write reads into separate files. Read
                          number will be suffixed to the file
                          name. NOTE! The `--split-3` option is
                          recommended. In cases where not all
                          spots have the same number of reads,
                          this option will produce files that WILL
                          CAUSE ERRORS in most programs which
                          process split pair fastq files.
--split-3               3-way splitting for mate-pairs. For each
                          spot, if there are two biological reads
                          satisfying filter conditions, the first
                          is placed in the `*_1.fastq` file, and
                          the second is placed in the `*_2.fastq`
                          file. If there is only one biological
                          read satisfying the filter conditions,
                          it is placed in the `*.fastq` file.All
                          other reads in the spot are ignored.
</code></pre><p>Okay, now it seems they recommend the <code>--split-3</code> option, so I guess I will use <code>fasterq-dump</code> with the <code>--split-3</code> flag, which is what I do on a regular basis due to the recommendation in the help message. However, according to the help message, only <code>*_1.fastq</code> and <code>*_2.fastq</code> files will be kept, and all other reads are ignored. In this case, &ldquo;other reads&rdquo; means &ldquo;technical reads&rdquo; which actually contain the index reads (cell barcodes) we want. Luckily, <code>fasterq-dump</code> has a <code>--include-technical</code> option. Thank Jon Trow from NLM Support helpdesk for letting me know this option. This should be good for us. Now let&rsquo;s do it using <code>SRR9672090</code> from <a href=https://www.nature.com/articles/s41587-019-0290-0><strong>SNARE-seq</strong></a> as an example, which is the ATAC modality from the mouse cortex experiments:</p><pre tabindex=0><code>$ fasterq-dump --split-3 --include-technical SRR9672090
spots read      : 124,642,473
reads read      : 373,927,419
reads written   : 249,284,946
technical reads : 124,642,473
</code></pre><p>However, only two files, <code>SRR9672090_1.fastq</code> and <code>SRR9672090_2.fastq</code>, were generated, each with 124,642,473 reads. I don&rsquo;t know where the index reads (technical reads) are. Maybe I should try the not-recommended <code>--split-files</code> option:</p><pre tabindex=0><code>$ fasterq-dump --split-files --include-technical SRR9672090
spots read      : 124,642,473
reads read      : 373,927,419
reads written   : 373,927,419
</code></pre><p>This time, three files <code>SRR9672090_1.fastq</code>, <code>SRR9672090_2.fastq</code> and <code>SRR9672090_3.fastq</code>, were generated, each with 124,642,473 reads. Great! This is what I want. The <code>SRR9672090_1.fastq</code> and <code>SRR9672090_2.fastq</code> are the actual ATAC fragment reads, and the <code>SRR9672090_3.fastq</code> contains the cell barcodes.</p><p>For the SNARE-seq data, now I actually learnt from Tim Stuart&rsquo;s <a href=https://timoast.github.io/blog/sra-aws/><strong>blog post</strong></a> and <a href=https://github.com/timoast/SNARE-seq><strong>github repository</strong></a> that you can directly download the user submitted fastq files. However, I cannot download via the URLs (due to geographical resitrctions ???).</p><p>Anyway, now let&rsquo;s try <code>SRR1947692</code>, which is the HEK293T and GM12878 mixture experiment from <a href=https://www.science.org/doi/10.1126/science.aab1601><strong>sci-ATAC-seq</strong></a>:</p><pre tabindex=0><code>$ fasterq-dump --split-files --include-technical SRR1947692
spots read      : 20,027,825
reads read      : 40,055,650
reads written   : 40,055,650
</code></pre><p>Sadly, only two files <code>SRR1947692_1.fastq</code> and <code>SRR1947692_2.fastq</code> were generated, each with 20,027,825 reads. This is kind of expected, because in the early days of single cell genomics, it was a common pratice to put cell barcodes into the fastq heeader, and not to inlcude them as separate fastq files. Now let&rsquo;s just look at the fastq header:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ awk <span style=color:#98c379>&#39;NR%4==1&#39;</span> SRR1947692_1.fastq | head
</span></span><span style=display:flex><span>@SRR1947692.1 SHEN-MISEQ02:1:1101:15311:1341 <span style=color:#e06c75>length</span><span style=color:#56b6c2>=</span><span style=color:#d19a66>51</span>
</span></span><span style=display:flex><span>@SRR1947692.2 SHEN-MISEQ02:1:1101:16462:1350 <span style=color:#e06c75>length</span><span style=color:#56b6c2>=</span><span style=color:#d19a66>51</span>
</span></span><span style=display:flex><span>@SRR1947692.3 SHEN-MISEQ02:1:1101:16204:1351 <span style=color:#e06c75>length</span><span style=color:#56b6c2>=</span><span style=color:#d19a66>51</span>
</span></span><span style=display:flex><span>@SRR1947692.4 SHEN-MISEQ02:1:1101:16257:1355 <span style=color:#e06c75>length</span><span style=color:#56b6c2>=</span><span style=color:#d19a66>51</span>
</span></span><span style=display:flex><span>@SRR1947692.5 SHEN-MISEQ02:1:1101:15016:1376 <span style=color:#e06c75>length</span><span style=color:#56b6c2>=</span><span style=color:#d19a66>51</span>
</span></span></code></pre></div><p>Oh well &mldr; not so useful, right? Then, I saw this documentation from the <a href=https://baigal628.github.io/MAESTRO_documentation/sci_atac.html><strong>MAESTRO website</strong></a>, where they used the following command to get the fastq files with proper headers:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>fastq-dump --split-files --origfmt --defline-seq <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR1947692
</span></span></code></pre></div><p>It seems you could preserve the fastq header information in its originally submitted format with the <code>--origfmt</code> option, and use <code>--defline-seq</code> to specify the exact format. The <code>$sg</code> is the <code>spot-group</code>, which often contains the index information. Then, I think I can just use <code>fasterq-dump</code> for faster processing. Since I knew that there are no technical reads in this case, so I did not include that flag. However, I got an error message:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fasterq-dump --split-files --origfmt --defline-seq <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR1947692
</span></span><span style=display:flex><span>unrecognized option: <span style=color:#98c379>&#39;--origfmt&#39;</span>
</span></span></code></pre></div><p>It turns out that the option <code>--origfmt</code> <a href=https://github.com/ncbi/sra-tools/issues/233><strong>is not supported anymore</strong></a> in <code>fasterq-dump</code>. Let&rsquo;s remove it to see what happens:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fasterq-dump --split-files --defline-seq <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR1947692
</span></span><span style=display:flex><span>unrecognized option: <span style=color:#98c379>&#39;--defline-seq&#39;</span>
</span></span></code></pre></div><p>That was just brilliant &mldr; By invoking the <code>fasterq-dump</code> help message, we know that we should use <code>--seq-defline</code> instead:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fasterq-dump --split-files --seq-defline <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR1947692
</span></span><span style=display:flex><span>spots <span style=color:#e5c07b>read</span>      : 20,027,825
</span></span><span style=display:flex><span>reads <span style=color:#e5c07b>read</span>      : 40,055,650
</span></span><span style=display:flex><span>reads written   : 40,055,650
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>$ awk <span style=color:#98c379>&#39;NR%4==1&#39;</span> SRR1947692_1.fastq | head -5
</span></span><span style=display:flex><span>@rd.1::SHEN-MISEQ02:1:1101:15311:1341
</span></span><span style=display:flex><span>@rd.2::SHEN-MISEQ02:1:1101:16462:1350
</span></span><span style=display:flex><span>@rd.3::SHEN-MISEQ02:1:1101:16204:1351
</span></span><span style=display:flex><span>@rd.4::SHEN-MISEQ02:1:1101:16257:1355
</span></span><span style=display:flex><span>@rd.5::SHEN-MISEQ02:1:1101:15016:1376
</span></span></code></pre></div><p>Oh, it seems without the <code>--origfmt</code>, <code>$sg</code> will not be written in the output (Note there is no content between the first and second colons in the header). I should have just used the exact the same command in the MAESTRO website:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fastq-dump --split-files --origfmt --defline-seq <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR1947692
</span></span><span style=display:flex><span>Read <span style=color:#d19a66>20027825</span> spots <span style=color:#c678dd>for</span> SRR1947692
</span></span><span style=display:flex><span>Written <span style=color:#d19a66>20027825</span> spots <span style=color:#c678dd>for</span> SRR1947692
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>$ awk <span style=color:#98c379>&#39;NR%4==1&#39;</span> SRR1947692_1.fastq | head -5
</span></span><span style=display:flex><span>@rd.1:TCTCCCGCCGAGGCTGACTGCATAAGGCGAAT:SHEN-MISEQ02:1:1101:15311:1341
</span></span><span style=display:flex><span>@rd.2:CGCGTTCCAGGCCGTACCTGCATAAGGCGACG:SHEN-MISEQ02:1:1101:16462:1350
</span></span><span style=display:flex><span>@rd.3:GAGATTCCCGTACTAGAAGGAGTATATAGCCT:SHEN-MISEQ02:1:1101:16204:1351
</span></span><span style=display:flex><span>@rd.4:GAGATTCCCGTACTAGGTAAGGAGATAGAGGC:SHEN-MISEQ02:1:1101:16257:1355
</span></span><span style=display:flex><span>@rd.5:ATTACTCGAAGAGGCACGAGTAGACAGGACGT:SHEN-MISEQ02:1:1101:15016:1376
</span></span></code></pre></div><p>Ha, I&rsquo;m so happy &mldr; You see the 32-bp DNA sequences between the first and the second colons in the fastq header? Those are the cell barcodes, each of which is a combination of <code>Tn5-T5</code> (8 bp), <code>Tn5-T7</code> (8 bp), <code>Nextera i5</code> (8 bp) and <code>Nextera i7</code> (8 bp). Check <a href=https://teichlab.github.io/scg_lib_structs/methods_html/sci-ATAC-seq.html><strong>this page</strong></a> for more information.</p><p>Finally, for the data that DO have technical reads available like SNARE-seq, what if you want both the technical reads and the <code>$sg</code> information? Well &mldr; I&rsquo;m afraid that you cannot use <code>fasterq-dump</code>. You have to do this in a slower way using <code>fastq-dump</code>. The following command will give you three fastq files, with the sample index in the fastq header.</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fastq-dump --split-files --origfmt --defline-seq <span style=color:#98c379>&#39;@rd.$si:$sg:$sn&#39;</span> SRR9672090
</span></span><span style=display:flex><span>Read <span style=color:#d19a66>124642473</span> spots <span style=color:#c678dd>for</span> SRR9672090
</span></span><span style=display:flex><span>Written <span style=color:#d19a66>124642473</span> spots <span style=color:#c678dd>for</span> SRR9672090
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span>$ awk <span style=color:#98c379>&#39;NR%4==1&#39;</span> SRR9672090_3.fastq | head -5
</span></span><span style=display:flex><span>@rd.1:TATCCTCT:D00611:697:CD0V6ANXX:5:2301:1176:2478
</span></span><span style=display:flex><span>@rd.2:TATCCTCT:D00611:697:CD0V6ANXX:5:2301:1480:2408
</span></span><span style=display:flex><span>@rd.3:TATCCTAT:D00611:697:CD0V6ANXX:5:2301:1361:2447
</span></span><span style=display:flex><span>@rd.4:TATCCTCT:D00611:697:CD0V6ANXX:5:2301:1384:2495
</span></span><span style=display:flex><span>@rd.5:TATCCTCT:D00611:697:CD0V6ANXX:5:2301:1335:2498
</span></span></code></pre></div><p>In this case, you don&rsquo;t need to specify <code>--include-technical</code> because it will not work with <code>fastq-dump</code>. The <code>--split-files</code> flag will dump technical reads if there is any. Anyway, I hope this is useful to people who do not use sratools very often, like me.</p></p></div><div class=prev-next></div></div></main><footer class=footer><span>&copy; 2025 Not A Rocket Scientist</span>
<span>Made with &#10084;&#65039; using <a target=_blank href=https://github.com/526avijitgupta/gokarna>Gokarna</a></span></footer></body></html>