<!doctype html><html lang=en><head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><link rel=apple-touch-icon sizes=180x180 href=/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/favicon-16x16.png><link rel=manifest href=/site.webmanifest><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><style>:root{--accent-color:#FF4D4D}</style><title>PIPseq$^{\textmd{TM}}$ V4 Update</title><meta name=description content="Genomics | Bioinformatics | Epigenetics"><meta name=keywords content='blog,gokarna,hugo,data science,biology,genomics,single cell genomics,sequencing library,illumina sequencing,fluentbio,pipseq,seqspec'><meta property="og:url" content="http://localhost:1313/posts/2023-07-02-pip-seq-v4-update/"><meta property="og:type" content="website"><meta property="og:title" content="PIPseq$^{\textmd{TM}}$ V4 Update"><meta property="og:description" content="Genomics | Bioinformatics | Epigenetics"><meta property="og:image" content="/images/profile_avatar.png"><meta name=twitter:card content="summary_large_image"><meta name=twitter:title content="PIPseq$^{\textmd{TM}}$ V4 Update"><meta name=twitter:description content="Genomics | Bioinformatics | Epigenetics"><meta property="twitter:domain" content="http://localhost:1313/posts/2023-07-02-pip-seq-v4-update/"><meta property="twitter:url" content="http://localhost:1313/posts/2023-07-02-pip-seq-v4-update/"><meta name=twitter:image content="/images/profile_avatar.png"><link rel=canonical href=http://localhost:1313/posts/2023-07-02-pip-seq-v4-update/><link rel=stylesheet type=text/css href=/css/normalize.min.css media=print onload='this.media="all"'><link rel=stylesheet type=text/css href=/css/main.css><link disabled id=dark-theme rel=stylesheet href=/css/dark.css><script src=/js/svg-injector.min.js></script><script src=/js/feather-icons.min.js></script><script src=/js/main.js></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css integrity=sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ crossorigin=anonymous><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js integrity=sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB crossorigin=anonymous></script><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js integrity=sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl crossorigin=anonymous onload=renderMathInElement(document.body)></script><script>document.addEventListener("DOMContentLoaded",function(){renderMathInElement(document.body,{delimiters:[{left:"$$",right:"$$",display:!0},{left:"$",right:"$",display:!1}]})})</script></head><body><script type=text/javascript>setThemeByUserPref()</script><header class=header><nav class=header-nav><div class=avatar><a href=http://localhost:1313/><img src=http://localhost:1313//images/profile_avatar.png alt=avatar></a></div><div class=nav-title><a class=nav-brand href=http://localhost:1313/>Not A Rocket Scientist</a></div><div class=nav-links><div class=nav-link><a href=http://localhost:1313/><span data-feather=home></span> Home</a></div><div class=nav-link><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></div><div class=nav-link><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></div><div class=nav-link><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></div><div class=nav-link><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></div><div class=nav-link><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></div><div class=nav-link><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></div><span class=nav-icons-divider></span><div class="nav-link dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only></span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></div><div class=nav-link id=hamburger-menu-toggle><span id=hamburger-menu-toggle-screen-reader-target class=sr-only>menu</span>
<a><span data-feather=menu></span></a></div><ul class="nav-hamburger-list visibility-hidden"><li class=nav-item><a href=http://localhost:1313/><span data-feather=home></span> Home</a></li><li class=nav-item><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></li><li class=nav-item><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></li><li class=nav-item><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></li><li class=nav-item><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></li><li class=nav-item><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></li><li class=nav-item><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></li><li class="nav-item dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only>theme</span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></li></ul></div></nav></header><main id=content><div class="post container"><div class=post-header-section><h1>PIPseq$^{\textmd{TM}}$ V4 Update</h1><small role=doc-subtitle></small><p class=post-date>July 2, 2023</p><ul class=post-tags><li class=post-tag><a href=http://localhost:1313/tags/single-cell-genomics>single cell genomics</a></li><li class=post-tag><a href=http://localhost:1313/tags/sequencing-library>sequencing library</a></li><li class=post-tag><a href=http://localhost:1313/tags/illumina-sequencing>illumina sequencing</a></li><li class=post-tag><a href=http://localhost:1313/tags/fluentbio>fluentbio</a></li><li class=post-tag><a href=http://localhost:1313/tags/pipseq>pipseq</a></li><li class=post-tag><a href=http://localhost:1313/tags/seqspec>seqspec</a></li></ul></div><div class=post-content><p><p><strong>NOTE:</strong> this post is just a slightly more detailed version of this <a href=https://twitter.com/XiChenUoM/status/1671400417350017025>Twitter thread</a>.</p><p>Not very long after I wrote <a href=/posts/2023-05-28-guessing-library-structures-1/>the series</a> about guessing the library layout of PIP-seq, the raw data (FASTQ files) for the V4 chemistry were made available from the <a href=https://www.fluentbio.com/datasets/>FluentBio website</a>, which is really nice.</p><p>Now let&rsquo;s have a look at what has been changed. The experimental procedures seem to remain the same. What about the library layout and barcodes? Let&rsquo;s use the <strong>T20 Kit V4 Barnyard Study</strong> data to figure that out. With the experience from the previous versions, we can immediately start with FastQC of Read 1.</p><p>This is the result of <code>SC20230206_FR_1_S1_L001_R1_001.fastq.gz</code>:</p><p><img src=/images/2023-07-02/fastqc.png alt></p><p>We do not see any common sequences in the middle of the read, like those in the V2 and V3. If we assume that the barcodes are still generated by the split-pool method, it often means there are some variable bases near the beginning part of the read to &ldquo;de-sync&rdquo; the sequencing, increasing the base complexity in each sequencing cycle. Another hint can be found in the last few bases of the read. We see that the dominant base is <code>T</code>, so this is basically the oligo-dT. However, unlike <a href=/posts/2023-06-04-guessing-library-structures-3/>the V2 chemistry</a> where almost 100% of the read have <code>T</code> at the last few bases, we only have ~80% of <code>T</code> at most. This also indicates there are some variable bases near the start of the read.</p><p>In this case, we need to do some manual alignment using the trick we mentioned previously. Get the first 40 high-quality sequences ignoring reads containing <code>N</code> by:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>zcat SC20230206_FR_1_S1_L001_R1_001.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    head -40
</span></span></code></pre></div><p>Put the output into VS Code. Bear in mind that there should be some common sequences around the middle part of the read, we can start manually aligning them like this:</p><p><img src=/images/2023-07-02/v4top40.png alt></p><p>Now it is clear to us the read structure is: <code>None/T/GT/TGA</code> + <code>8bp bc1</code> + <code>ATG</code> + <code>6bp bc2</code> + <code>CAG</code> + <code>6bp bc3</code> + <code>TCGAG</code> + <code>8bp bc4</code> + <code>12 bp UMI</code> + <code>T</code>. That is almost the same as the <a href=/posts/2023-06-03-guessing-library-structures-2/>V3 chemistry</a>, except the variable bases <code>None/T/GT/TGA</code> at the beginning. Therefore, we have reasons to believe that the whitelist remains unchanged.</p><p>Is this the case? Let&rsquo;s find out by processing the data using the V3 whitelist. Since we know the read structure, this can be done by <code>STARsolo</code> using the following command:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>STAR --runThreadN <span style=color:#d19a66>40</span> <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --genomeDir reference/others/hs_mm_mix/star_2.7.9a <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --readFilesCommand zcat <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outFileNamePrefix ./star_outs/ <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --readFilesIn SC20230206_FR_1_S1_L001_R2_001.fastq.gz <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>                   SC20230206_FR_1_S1_L001_R1_001.fastq.gz <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloType CB_UMI_Complex <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloBarcodeReadLength <span style=color:#d19a66>0</span> <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloAdapterSequence ATGNNNNNNGAGNNNNNNTCGAG <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBposition 2_-8_2_-1 2_3_2_8 2_12_2_17 2_23_2_30 <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloUMIposition 2_31_2_42 <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBwhitelist fb_v3_bc1.tsv fb_v3_bc2.tsv fb_v3_bc3.tsv fb_v3_bc4.tsv <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBmatchWLtype 1MM <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCellFilter EmptyDrops_CR <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outSAMattributes CB UB <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outSAMtype BAM SortedByCoordinate
</span></span></code></pre></div><p>Once it is done, we can check the content of the <code>Summary.csv</code>:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code class=language-bash data-lang=bash><span style=display:flex><span>Number of Reads,1067049515
</span></span><span style=display:flex;background-color:#3d4148><span>Reads With Valid Barcodes,0.953267
</span></span><span style=display:flex><span>Sequencing Saturation,0.397782
</span></span><span style=display:flex><span>Q30 Bases in CB+UMI,0.964693
</span></span><span style=display:flex><span>Q30 Bases in RNA read,0.935793
</span></span><span style=display:flex><span>Reads Mapped to Genome: Unique+Multiple,0.857187
</span></span><span style=display:flex><span>Reads Mapped to Genome: Unique,0.648044
</span></span><span style=display:flex><span>Reads Mapped to Gene: Unique+Multipe Gene,0.719135
</span></span><span style=display:flex><span>Reads Mapped to Gene: Unique Gene,0.692885
</span></span><span style=display:flex><span>Estimated Number of Cells,40579
</span></span><span style=display:flex><span>Unique Reads in Cells Mapped to Gene,677629098
</span></span><span style=display:flex><span>Fraction of Unique Reads in Cells,0.916529
</span></span><span style=display:flex><span>Mean Reads per Cell,16699
</span></span><span style=display:flex><span>Median Reads per Cell,15482
</span></span><span style=display:flex><span>UMIs in Cells,407207180
</span></span><span style=display:flex><span>Mean UMI per Cell,10034
</span></span><span style=display:flex><span>Median UMI per Cell,9348
</span></span><span style=display:flex><span>Mean Gene per Cell,2571
</span></span><span style=display:flex><span>Median Gene per Cell,2581
</span></span><span style=display:flex><span>Total Gene Detected,43477</span></span></code></pre></div><p>We see that more than 95% of the barcodes are valid, which confirms our guess.</p><p>Another sanity check to do is whether the first base of Read 2 is all <code>T</code>. We expect this to be the case since the library preparation procedures are almost the same as V3. The Read 2 adapter is added through fragmentation, A tailing and ligation. Therefore, all Read 2 should start with a <code>T</code>. Now let&rsquo;s have a look:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ zcat SC20230206_FR_1_S1_L001_R2_001.fastq.gz | sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | cut -c <span style=color:#d19a66>1</span> | sort | uniq -c
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#d19a66>286396621</span> A
</span></span><span style=display:flex><span><span style=color:#d19a66>129912129</span> C
</span></span><span style=display:flex><span><span style=color:#d19a66>264907938</span> G
</span></span><span style=display:flex><span><span style=color:#d19a66>385832827</span> T
</span></span></code></pre></div><p>It is not what we expected. Maybe the first cycle of Read 2 is a dark cycle. I&rsquo;m not entirely sure.</p></p></div><div class=prev-next></div></div><aside class=post-toc><nav id=toc><nav id=TableOfContents></nav></nav></aside></main><footer class=footer><span>&copy; 2025 Not A Rocket Scientist</span>
<span>Made with &#10084;&#65039; using <a target=_blank href=https://github.com/526avijitgupta/gokarna>Gokarna</a></span></footer></body></html>