<!doctype html><html lang=en><head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><link rel=apple-touch-icon sizes=180x180 href=/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/favicon-16x16.png><link rel=manifest href=/site.webmanifest><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><style>:root{--accent-color:#FF4D4D}</style><title>Guessing Library Structures (3) - PIP-seq V2</title><meta name=description content="Genomics | Bioinformatics | Epigenetics"><meta name=keywords content='blog,gokarna,hugo,data science,biology,genomics,single cell genomics,sequencing library,illumina sequencing,fluentbio,pipseq,seqspec'><meta property="og:url" content="http://localhost:1313/posts/2023-06-04-guessing-library-structures-3/"><meta property="og:type" content="website"><meta property="og:title" content="Guessing Library Structures (3) - PIP-seq V2"><meta property="og:description" content="Genomics | Bioinformatics | Epigenetics"><meta property="og:image" content="/images/profile_avatar.png"><meta name=twitter:card content="summary_large_image"><meta name=twitter:title content="Guessing Library Structures (3) - PIP-seq V2"><meta name=twitter:description content="Genomics | Bioinformatics | Epigenetics"><meta property="twitter:domain" content="http://localhost:1313/posts/2023-06-04-guessing-library-structures-3/"><meta property="twitter:url" content="http://localhost:1313/posts/2023-06-04-guessing-library-structures-3/"><meta name=twitter:image content="/images/profile_avatar.png"><link rel=canonical href=http://localhost:1313/posts/2023-06-04-guessing-library-structures-3/><link rel=stylesheet type=text/css href=/css/normalize.min.css media=print onload='this.media="all"'><link rel=stylesheet type=text/css href=/css/main.css><link disabled id=dark-theme rel=stylesheet href=/css/dark.css><script src=/js/svg-injector.min.js></script><script src=/js/feather-icons.min.js></script><script src=/js/main.js></script><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css integrity=sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ crossorigin=anonymous><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js integrity=sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB crossorigin=anonymous></script><script defer src=https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js integrity=sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl crossorigin=anonymous onload=renderMathInElement(document.body)></script><script>document.addEventListener("DOMContentLoaded",function(){renderMathInElement(document.body,{delimiters:[{left:"$$",right:"$$",display:!0},{left:"$",right:"$",display:!1}]})})</script></head><body><script type=text/javascript>setThemeByUserPref()</script><header class=header><nav class=header-nav><div class=avatar><a href=http://localhost:1313/><img src=http://localhost:1313//images/profile_avatar.png alt=avatar></a></div><div class=nav-title><a class=nav-brand href=http://localhost:1313/>Not A Rocket Scientist</a></div><div class=nav-links><div class=nav-link><a href=http://localhost:1313/><span data-feather=home></span> Home</a></div><div class=nav-link><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></div><div class=nav-link><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></div><div class=nav-link><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></div><div class=nav-link><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></div><div class=nav-link><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></div><div class=nav-link><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></div><span class=nav-icons-divider></span><div class="nav-link dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only></span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></div><div class=nav-link id=hamburger-menu-toggle><span id=hamburger-menu-toggle-screen-reader-target class=sr-only>menu</span>
<a><span data-feather=menu></span></a></div><ul class="nav-hamburger-list visibility-hidden"><li class=nav-item><a href=http://localhost:1313/><span data-feather=home></span> Home</a></li><li class=nav-item><a href=http://localhost:1313/about/><span data-feather=user></span> About</a></li><li class=nav-item><a href=http://localhost:1313/posts/><span data-feather=list></span> Posts</a></li><li class=nav-item><a href=http://localhost:1313/projects/><span data-feather=square></span> Projects</a></li><li class=nav-item><a href=http://localhost:1313/tags/><span data-feather=tag></span> Tags</a></li><li class=nav-item><a href=https://dbrg77.github.io/SUSTech-BIO210><span data-feather=bar-chart-2></span></a></li><li class=nav-item><a href=https://github.com/Teichlab/scg_lib_structs><span data-feather=smile></span></a></li><li class="nav-item dark-theme-toggle"><span id=dark-theme-toggle-screen-reader-target class=sr-only>theme</span>
<a><span id=theme-toggle-icon data-feather=moon></span></a></li></ul></div></nav></header><main id=content><div class="post container"><div class=post-header-section><h1>Guessing Library Structures (3) - PIP-seq V2</h1><small role=doc-subtitle></small><p class=post-date>June 4, 2023</p><ul class=post-tags><li class=post-tag><a href=http://localhost:1313/tags/single-cell-genomics>single cell genomics</a></li><li class=post-tag><a href=http://localhost:1313/tags/sequencing-library>sequencing library</a></li><li class=post-tag><a href=http://localhost:1313/tags/illumina-sequencing>illumina sequencing</a></li><li class=post-tag><a href=http://localhost:1313/tags/fluentbio>fluentbio</a></li><li class=post-tag><a href=http://localhost:1313/tags/pipseq>pipseq</a></li><li class=post-tag><a href=http://localhost:1313/tags/seqspec>seqspec</a></li></ul></div><div class=post-content><p><div class="admonition note"><div class=title>Portal</div><div class=content><ul><li><a href=/posts/2023-05-28-guessing-library-structures-1/>Guessing Library Structures (1) - Background</a></li><li><a href=/posts/2023-06-03-guessing-library-structures-2/>Guessing Library Structures (2) - PIPseq$^{\textmd{TM}}$ V3</a></li><li><a href=/posts/2023-06-04-guessing-library-structures-3/>Guessing Library Structures (3) - PIP-seq V2</a></li><li><a href=/posts/2023-06-07-guessing-library-structures-4/>Guessing Library Structures (4) - PIP-seq V1prototype</a></li></ul></div></div><p>This is the third post from a series where I use the recent <a href=https://www.nature.com/articles/s41587-023-01685-z>PIP-seq</a> data as an example to demonstrate the importance to use a common standard like <a href=https://www.biorxiv.org/content/10.1101/2023.03.17.533215v1><em>seqspec</em></a> to accompany sequencing reads submissions into public repositories.</p><h2 id=some-notes>Some Notes</h2><p>In the previous post, we have figured out the library structure of many samples from the <strong>Clark2023</strong> paper. They are using the commerical <strong>FluentBio</strong> kits, so they basically have the <a href=https://teichlab.github.io/scg_lib_structs/methods_html/PIP-seq.html>PIPseq$^{\textmd{TM}}$ V3 library structure</a>. However, we still haven&rsquo;t figure out what is going on with samples like <code>SRR19086119</code>. We will tackle this in the next post.</p><p>On top of <code>SRR19086119</code> and the like, there is still another version of the method. Check the <strong>&ldquo;GSM6138415 Mixed species&rdquo;</strong> sample, which eventually leads to <code>SRR19180490</code>. We can download the FASTQ files:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>$ fastq-dump --split-3 -A SRR19180490
</span></span></code></pre></div><p><a href=https://www.fluentbio.com/products/pipseeker-software-for-data-analysis/>PIPseeker$^{\textmd{TM}}$</a> cannot recognise the FASTQ files, so it is neither <code>v3</code> nor <code>v4</code>. I&rsquo;m pretty sure it is not used in the commercial version, but it was used in the Clark2023 paper. Therefore, I call this <strong>PIP-seq V2</strong>. This is just me giving it a name, not necessarily correct.</p><h2 id=srr19180490>SRR19180490</h2><p>We already had some previous experience, so we could immediately check Read 1 with FastQC:</p><p><img src=/images/2023-06-04/pip-seq_srr19180490.png alt></p><p>It is immediately clear to us that there are some constant regions (linkers or splints), so it is reasonable to guess the variable regions flanking the linkers are barcodes or UMIs:</p><table><thead><tr><th style=text-align:center>Base positions</th><th style=text-align:center>Identity</th></tr></thead><tbody><tr><td style=text-align:center>1 - 8</td><td style=text-align:center>a variable region (possibly 8 bp barcode1)</td></tr><tr><td style=text-align:center>9 - 15</td><td style=text-align:center>ATGCATS (Here, &ldquo;S&rdquo; is ~65% C and ~35% G)</td></tr><tr><td style=text-align:center>16 - 23</td><td style=text-align:center>a variable region (possibly 8 bp barcode2)</td></tr><tr><td style=text-align:center>24 - 30</td><td style=text-align:center>SCTCGAG (Here, &ldquo;S&rdquo; is ~75% C and ~25% G)</td></tr><tr><td style=text-align:center>31 - 38</td><td style=text-align:center>a variable region (possibly 8 bp barcode3)</td></tr><tr><td style=text-align:center>39 - 50</td><td style=text-align:center>a variable region (possibly 12 bp UMI)</td></tr><tr><td style=text-align:center>51 - 55</td><td style=text-align:center>TTTTT (this is from oligo-dT)</td></tr></tbody></table><p>In the constant regions, there is a base with S (G or C). Not sure why this happens, but we are going to just use the most frequent base to represent them. That is, <code>ATGCATS</code> is <code>ATGCATC</code>, and <code>SCTCGAG</code> is <code>CCTCGAG</code>. It does not affect the results.</p><p>Naturally, we would guess there should be fixed sequences in those three variable regions. Let&rsquo;s find out using the similar technique from the previous post:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>zcat SRR19180490_1.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    cut -c 1-8 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort | uniq -c |
</span></span><span style=display:flex><span>    sort -b -k1,1nr | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    less -N
</span></span></code></pre></div><p>The output is all unique 8-mers in the first 8 bases for Read 1, sorted by occurrences in descending order. As expected, the occurrences of the top 96 8-mers are within a similar range, but there is a sudden an order of magnitude drop from the 96th to the 97th 8-mers. Therefore, we are relatively confident that the top 96 8-mers are in our whitelist. Here are the lines 94 - 98 of the output:</p><div class=highlight><div style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><table style=border-spacing:0;padding:0;margin:0;border:0><tr><td style=vertical-align:top;padding:0;margin:0;border:0><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">94
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">95
</span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">96
</span></span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">97
</span></span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">98
</span></code></pre></td><td style=vertical-align:top;padding:0;margin:0;border:0;width:100%><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code class=language-bash data-lang=bash><span style=display:flex><span><span style=color:#d19a66>811868</span> ATTACCTT
</span></span><span style=display:flex><span><span style=color:#d19a66>757276</span> TATGTGAA
</span></span><span style=display:flex;background-color:#3d4148><span><span style=color:#d19a66>686049</span> TATCCACG
</span></span><span style=display:flex;background-color:#3d4148><span> <span style=color:#d19a66>19354</span> AAAAAACA
</span></span><span style=display:flex><span> <span style=color:#d19a66>16786</span> AAAAAAAA</span></span></code></pre></td></tr></table></div></div><p>Similarly, we could check bases 16 - 23:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>zcat SRR19180490_1.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    cut -c 16-23 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort | uniq -c |
</span></span><span style=display:flex><span>    sort -b -k1,1nr | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    less -N
</span></span></code></pre></div><p>Lines 94 - 102 from the output are shown here:</p><div class=highlight><div style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><table style=border-spacing:0;padding:0;margin:0;border:0><tr><td style=vertical-align:top;padding:0;margin:0;border:0><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 94
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 95
</span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 96
</span></span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 97
</span></span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 98
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f"> 99
</span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">100
</span></span><span style=background-color:#3d4148><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">101
</span></span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:.4em;padding:0 .4em;color:#55595f">102
</span></code></pre></td><td style=vertical-align:top;padding:0;margin:0;border:0;width:100%><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code class=language-bash data-lang=bash><span style=display:flex><span><span style=color:#d19a66>896729</span> TGGAAAGC
</span></span><span style=display:flex><span><span style=color:#d19a66>864565</span> CTTAGGCC
</span></span><span style=display:flex;background-color:#3d4148><span><span style=color:#d19a66>603053</span> ATCCCATT
</span></span><span style=display:flex;background-color:#3d4148><span><span style=color:#d19a66>389631</span> CCCATTCA
</span></span><span style=display:flex><span><span style=color:#d19a66>275130</span> CTGCGTTG
</span></span><span style=display:flex><span><span style=color:#d19a66>155961</span> CTCCCATT
</span></span><span style=display:flex;background-color:#3d4148><span><span style=color:#d19a66>152746</span> TTCCCATT
</span></span><span style=display:flex;background-color:#3d4148><span> <span style=color:#d19a66>59805</span> CCCCATTC
</span></span><span style=display:flex><span> <span style=color:#d19a66>40335</span> TCCCATTC</span></span></code></pre></td></tr></table></div></div><p>In this case, the differences of occurrences among 96th - 100th 8-mers are not very different. There is a sudden an order of magnitude drop in frequency from 100th to 101st 8-mer. What&rsquo;s going on with the 97th - 100th 8-mers? If we check carefully enough, we could find out that each of them has only 1 bp variation compared to some of the 8-mers from the top 96. Therefore, it seems we could still just use the top 96 as the whitelist.</p><p>Bases 31 - 38 are in the similar situation. Therefore, we could extract the whitelist by:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span><span style=color:#7f848e># get barcode1 (8 bp)</span>
</span></span><span style=display:flex><span>zcat SRR19180490_1.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    cut -c 1-8 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort | uniq -c | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort -b -k1,1nr | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    head -96 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    awk <span style=color:#98c379>&#39;{print $2}&#39;</span> &gt; pip-seq_v2_bc1.tsv
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#7f848e># get barcode2 (8 bp)</span>
</span></span><span style=display:flex><span>zcat SRR19180490_1.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    cut -c 16-23 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort | uniq -c | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort -b -k1,1nr | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    head -96 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    awk <span style=color:#98c379>&#39;{print $2}&#39;</span> &gt; pip-seq_v2_bc2.tsv
</span></span><span style=display:flex><span>
</span></span><span style=display:flex><span><span style=color:#7f848e># get barcode3 (8 bp)</span>
</span></span><span style=display:flex><span>zcat SRR19180490_1.fastq.gz | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sed -n <span style=color:#98c379>&#39;2~4 p&#39;</span> | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    cut -c 31-38 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort | uniq -c | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    sort -b -k1,1nr | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    head -96 | <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>    awk <span style=color:#98c379>&#39;{print $2}&#39;</span> &gt; pip-seq_v2_bc3.tsv
</span></span></code></pre></div><p>On top of that, by checking the method section of the Clark2023 paper, we finally figure out the library structure of PIP-seq V2. The HTML representation is <a href=https://teichlab.github.io/scg_lib_structs/methods_html/PIP-seq.html>here</a>, and <strong>the <em>seqspec</em> specification</strong> is <a href=https://github.com/IGVF/seqspec/tree/main/specs/PIPseqV2>here</a>.</p><h2 id=validation>Validation</h2><p>Once we got the whitelist for all three parts, we could start the validation using <a href=https://github.com/alexdobin/STAR>STARsolo</a>. The command is straightforward to understand after we figured out the library structure:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>STAR --runThreadN <span style=color:#d19a66>40</span> <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --genomeDir /data/bio-chenx/reference/others/hs_mm_mix/refdata-gex-GRCh38-and-mm10-2020-A/star_2.7.9a <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --readFilesCommand zcat <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outFileNamePrefix ./star_outs/SRR19180490/ <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --readFilesIn fastqs/SRR19180490_2.fastq.gz fastqs/SRR19180490_1.fastq.gz <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloType CB_UMI_Complex <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBposition 0_0_0_7 0_15_0_22 0_30_0_37 <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloUMIposition 0_38_0_49 <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloBarcodeReadLength <span style=color:#d19a66>0</span> <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBwhitelist pip-seq_v2_bc1.tsv pip-seq_v2_bc2.tsv pip-seq_v2_bc3.tsv <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCBmatchWLtype 1MM <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --soloCellFilter EmptyDrops_CR <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outSAMattributes CB UB <span style=color:#98c379>\
</span></span></span><span style=display:flex><span><span style=color:#98c379></span>     --outSAMtype BAM SortedByCoordinate
</span></span></code></pre></div><p>Check the result summary:</p><div class=highlight><pre tabindex=0 style=color:#abb2bf;background-color:#282c34;-moz-tab-size:4;-o-tab-size:4;tab-size:4;display:grid><code class=language-bash data-lang=bash><span style=display:flex><span>$ cat star_outs/SRR19180490/Solo.out/Gene/Summary.csv
</span></span><span style=display:flex><span>Number of Reads,129331399
</span></span><span style=display:flex;background-color:#3d4148><span>Reads With Valid Barcodes,0.934121
</span></span><span style=display:flex><span>Sequencing Saturation,0.360427
</span></span><span style=display:flex><span>Q30 Bases in CB+UMI,0.958061
</span></span><span style=display:flex><span>Q30 Bases in RNA read,0.943714
</span></span><span style=display:flex><span>Reads Mapped to Genome: Unique+Multiple,0.823944
</span></span><span style=display:flex><span>Reads Mapped to Genome: Unique,0.608917
</span></span><span style=display:flex><span>Reads Mapped to Gene: Unique+Multipe Gene,0.682908
</span></span><span style=display:flex><span>Reads Mapped to Gene: Unique Gene,0.656224
</span></span><span style=display:flex><span>Estimated Number of Cells,2932
</span></span><span style=display:flex><span>Unique Reads in Cells Mapped to Gene,52596524
</span></span><span style=display:flex><span>Fraction of Unique Reads in Cells,0.619727
</span></span><span style=display:flex><span>Mean Reads per Cell,17938
</span></span><span style=display:flex><span>Median Reads per Cell,14059
</span></span><span style=display:flex><span>UMIs in Cells,31166927
</span></span><span style=display:flex><span>Mean UMI per Cell,10629
</span></span><span style=display:flex><span>Median UMI per Cell,8632
</span></span><span style=display:flex><span>Mean Gene per Cell,2362
</span></span><span style=display:flex><span>Median Gene per Cell,2541
</span></span><span style=display:flex><span>Total Gene Detected,33440</span></span></code></pre></div><p>More than 93% reads have valid barcodes, I guess this is quite acceptable. After this, we should use another sample for an independent validation. At this moment of writing, I only found <code>SRR19180490</code> had this structure, but I was not searching very hard &mldr; Oh well &mldr;</p></p></div><div class=prev-next></div></div><aside class=post-toc><nav id=toc><nav id=TableOfContents><ul><li><a href=#some-notes>Some Notes</a></li><li><a href=#srr19180490>SRR19180490</a></li><li><a href=#validation>Validation</a></li></ul></nav></nav></aside></main><footer class=footer><span>&copy; 2025 Not A Rocket Scientist</span>
<span>Made with &#10084;&#65039; using <a target=_blank href=https://github.com/526avijitgupta/gokarna>Gokarna</a></span></footer></body></html>